library(gridExtra)
library(GGally)
#' @title Exposomex Link
urlhead = 'http://www.exposomex.cn:8080/'

#' @title Initialize StatTidy module
#' @description Initialize StatTidy module analysis. It can generate an R6 class object.
#' @usage InitStatTidy()
#' @param
#' @details It is designed to tidy the data for the target model analysis.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatTidy()
#'      FuncExit(PID = res$PID)
#' @author Bin Wang

InitStatTidy = function(){
  url = paste0(urlhead,'InitTidy')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}

#' @title Load data file for StatTidy module
#' @description Load data file for StatTidy module
#' @usage LoadStatTidy(PID, UseExample = "default", DataPath=NULL, VocaPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatTidy
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input directory of vocabulary file, e.g. "D:/test/eg_voca_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatTidy()
#'    res = LoadStatTidy(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

LoadStatTidy =function(PID,
                       UseExample = "default",
                       DataPath = NULL,
                       VocaPath = NULL){
  url = paste0(urlhead,'LoadTidy')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}

  res = httr::POST(url,
                   body = serialize(list(PID = PID,
                                         UseExample = UseExample,
                                         Expodata = exdata,
                                         Vocadata = vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Missing data imputation.
#' @description Missing data imputation.
#' @usage TidyTransImput(PID,OutPath = "default",Vars,Method)
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Vars Variables to be imputed. Available options include:
#'     "all.x", all exposure variables;
#'     Users can also choose available variables
#'     by selecting "Other" option, and copy the variables by clicking "Available vars".
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param Method Methods used for imputation. Available options include "lod" or "cart" methods.
#'     For "lod" method, limit of detection (LOD) should be included in the "Vocabulary" file.
#' @details
#' @return An R6 class object containing variable(s) with imputation.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res2 = TidyTransImput(PID=res$PID, Vars="all.x", Method="lod")
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyTransImput =function(PID,
                         OutPath = "default",
                         Vars,
                         Method){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TransImput')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransImputVars=Vars,
                               TransImputMethod=Method),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyTransImput'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyTransImput/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyTransImput/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Delete variables with low variance
#' @description Whether to delete variables with low variance. The default option is "yes".
#'     If skipped, it may result in failure to build models.
#' @usage TidyDelNearZeroVar(PID, OutPath = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @details
#' @return An R6 class object containing the variable(s) with acceptable variance.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res2 = TidyDelNearZeroVar(PID=res$PID)
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyDelNearZeroVar = function(PID,
                              OutPath = "default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DelLowVar')
  res =  httr::POST(url,
                    body = list(PID=PID),
                    encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/DelLowVar'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/DelLowVar/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/DelLowVar/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Delete variables with missing values
#' @description Whether to delete missing variables with low variance. The default option is "yes".
#'     If skipped, it may result in failure during modeling.
#' @usage TidyDelMiss(PID, OutPath = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @details
#' @return An R6 class object containing the variable(s) without missing values.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res1$Expo$Data
#'     res2 = TidyDelMiss(PID=res$PID)
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyDelMiss = function(PID,
                       OutPath = "default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url =paste0(urlhead,'DelMiss')
  res = httr::POST(url,
                   body = list(PID=PID),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyDelMiss'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyDelMiss/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyDelMiss/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )
  return(results$eSet)
}


#' @title Transform data type
#' @description Transform data type
#' @usage TidyTransType(PID,OutPath = "default",Vars,To)
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Vars Variables to be imputed. Available options include:
#'     "all.x", all exposure variables;
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param To chr. Indicate the type of the chosen variables to be transformed into.
#'     Available options include "integer", "numeric", "character", "factor", "logical", and "date".
#' @details
#' @return An R6 class object containing the variable(s) after transforming data type.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res1$Expo$Data
#'     res2 = TidyTransType(PID=res$PID, Vars = "X1,X2", To = "character")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyTransType = function(PID,
                         OutPath = "default",
                         Vars,
                         To){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TransType')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransTypeVars=Vars,
                               TransTypeTo=To),
                   encode = 'json')

  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyTransType'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyTransType/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyTransType/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Classify variables into various groups
#' @description Classify variables into various groups
#' @usage TidyTransClass(PID, OutPath = "default", Vars, LevelTo)
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Vars Variables to be imputed. Available options include:
#'     "all.x", all exposure variables;
#'     Users can also choose available variables
#'     by selecting "Other" option, and copy the variables by clicking "Available vars".
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param LevelTo The number of levels to convert variables to.
#' @details
#' @return An R6 class object containing the variable(s) after classifying data into various levels.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res1$Expo$Data
#'     res2 = TidyTransClass(PID=res$PID, Vars="X1", LevelTo="4")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyTransClass = function(PID,
                          OutPath = "default",
                          Vars,
                          LevelTo){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TransClass')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransClassVars=Vars,
                               TransClassLevelTo=LevelTo),
                   encode = 'json')

  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyTransClass'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyTransClass/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyTransClass/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Scale variables
#' @description Scale variables
#' @usage TidyTransScale(PID,OutPath = "default",Vars,Method,...)
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Vars Variables to be imputed. Available options include:
#'     "all.x", all exposure variables;
#'     Users can also choose available variables
#'     by selecting "Other" option, and copy the variables by clicking "Available vars".
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param Method chr. Scaling methods. Available options include "normal" and "range".
#' @param Direct chr. Direction to be transformed, Available options include "positive" and "negative".
#' @param RangeLow num. Lower limit for range method.
#' @param RangeUpper num. Upper limit for range method. It should be greater than the lower limit.
#' @details
#' @return An R6 class object containing the variable(s) after scaling data.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res1$Expo$Data
#'     res2 = TidyTransScale(PID=res$PID, Vars="all.x", Method="normal")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyTransScale = function(PID,
                          OutPath = "default",
                          Vars,
                          Method = "normal",
                          Direct="positive",
                          RangeLow="0",
                          RangeUpper="1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TransScale')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransScaleVars=Vars,
                               TransScaleMethod=Method,
                               TransScaleDirect=Direct,
                               TransScaleRangeLow=RangeLow,
                               TransScaleRangeUpper=RangeUpper),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyTransScale'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyTransScale/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyTransScale/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Transform variable distribution
#' @description Transform variable distribution
#' @usage TidyTransDistr(PID,OutPath = "default", Vars, Method)
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Vars Variables to be imputed. Available options include:
#'     "all.x", all exposure variables; Users can also choose available variables
#'     by selecting "Other" option, and copy the variables by clicking "Available vars".
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param Method
#' @details
#' @return An R6 class object containing the variable(s) after transforming distribution.
#' @export
#' @examples res = InitStatTidy()
#'     res1 = LoadStatTidy(PID=res$PID, UseExample="example#1")
#'     res1$Expo$Data
#'     res2 = TidyTransDistr(PID=res$PID, Vars="X6,X7", Method="log10")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

TidyTransDistr = function(PID,
                          OutPath = "default",
                          Vars,
                          Method){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TransDistr')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransDistrVars=Vars,
                               TransDistrMethod=Method),
                   encode = 'json')

  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyTransDistr'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyTransDistr/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyTransDistr/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}

#' @title Initialize StatDesc Module.
#' @description The first step to start StatDesc Module.
#' @usage InitStatDesc().
#' @return An R6 class object.
#' @export
#' @examples res = InitStatDesc()
#'     FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

InitStatDesc = function(){
  url = paste0(urlhead,'InitDesc')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data for StatDesc Module.
#' @description Upload  data file for StatDesc Module.
#' @usage LoadStatDesc(PID, UseExample = "default", DataPath = NULL, VocaPath = NULL).
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param UseExample chr. Whether uses example data for analyses, available option include "example#1"
#'    for using example data1 and "default" for using data.
#' @param DataPath chr. Input file directory, e.g. "D:/test/StatDesc_data.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input file directory, e.g. "D:/test/StatDesc_voca.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @return A list object containing imported data.
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

LoadStatDesc =function(PID,
                       UseExample="default",
                       DataPath=NULL,
                       VocaPath=NULL){
  url = paste0(urlhead,'LoadDesc')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)

    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata,
                                         Vocadata=vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Tidy data for StatDesc module
#' @description Tidy data file for StatDesc module
#' @usage TidyStatDesc(PID = PID)
#' @param PID chr. Program ID. It must be the same with the PID generated by
#'    InitStatDesc
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @details TidyStatDesc function resembles several tidy steps for data pre-treatment
#'    before executing other functions in StatDesc module, including deleting vars with
#'    low variance and deleting missing values.
#' @return An R6 class object containing the input data and vocabulary file.
#' @export
#' @examples res <- InitStatDesc()
#'    res1 <- LoadStatDesc(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 <- TidyStatDesc(PID = res$PID)
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang(corresponding author)

TidyStatDesc = function(PID,
                        OutPath = "default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyDesc')
  res = httr::POST(url,
                   body = list(PID=PID),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyStatDesc'))
  try(vroom::vroom_write(results$Expo$Data,
                         paste0(OutPath,'/TidyStatDesc/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$Expo$Voca,
                         paste0(OutPath,'/TidyStatDesc/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Create Table 1 for StatDesc Module.
#' @description Create Table 1 for different epidemilogical study designs.
#' @usage DescTable1(PID = res$PID,EpiDesign = "cohort", VarsY = "Y1", VarsC = "C1,C2,C3,C4,C5,C6",
#'    Missing = "ifany").
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param OutPath chr. Output file directory, e.g. "D:/StatDesc/DescTable1". It should be
#'    noted that the slash symbol is "/", not "\".If "default", the current working directory will be set.
#' @param EpiDesign chr. Research types provided for users, include "cohort" , "case.control", "cross.sectional".
#' @param VarsY chr. Outcome variable used for modelling. Only one variable can be entered.
#' @param VarsC chr. Covariate variables needing further statistical test.  It should be noted that there is
#'     fixed format for the entering characters separated with  "," and without space, e.g. "C1,C2,C3,C4,C5".
#'     The defaults value is all covariate variables listed in the data file, which can be entered with "all.c".
#' @param Missing chr. Counts of missing values in the table, available options include are "no" (never display missing values),
#'    "ifany" (only display if any missing values), and "always" (includes missing count row for all variables). Default is "ifany".
#' @return A list object containing standardized table 1
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    res2 =  TidyStatDesc(PID = res$PID)
#'    res3 = DescTable1(PID = res$PID, EpiDesign = "cohort", VarsY = "Y1", VarsC = "C1,C2,C3,C4,C5,C6", Missing = "ifany")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

DescTable1 = function(PID,
                      OutPath="default",
                      EpiDesign,
                      VarsY,
                      VarsC,
                      Missing){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DescTable1')
  res = httr::POST(url,
                   body = list(PID=PID,
                               DescTable1EpiDesign=EpiDesign,
                               DescTable1VarsY=VarsY,
                               DescTable1VarsC=VarsC,
                               DescTable1Missing=Missing),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/DescTable1'))
  for (name in names(results$Stat_Table1)){
    try(
      gt::gtsave(data=results$Stat_Table1[[name]],
                 filename = paste0(OutPath,'/DescTable1/',name)),
      silent = TRUE
    )
  }

  return(results)
}


#' @title Normality test.
#' @description Normality test for numeric variables.
#' @usage DescNorm(PID = res$PID, Vars = 'X5,X6,X7,X8,X9', Method = "shapiro.test", Layout = "rose.chart" ,
#'    Brightness = "light", Palette = "default3").
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param OutPath chr. Output file directory, e.g. "D:/StatDesc/DescNorm". It should be
#'    noted that the slash symbol is "/", not "\".If "default", the current working directory will be set.
#' @param Vars chr. Target variables used for modelling. It should be noted that there is fixed format for
#'    the entering characters separated with  "," and without space, e.g. "X7,X8,X9,X10,X11". The default
#'    values is "all" (all variables are included).
#' @param Method chr. Normality test method. Only "shapiro.test" method is available at present.
#' @param Layout chr. Visualization layout. Available values include "column", "column.points", "rose.chart", and "density".
#' @param Brightness chr. Visualization brightness.  Available values include "light" and "dark".
#' @param Palette chr. Visualization palette. Available values include "default1", "default2" and 6 journal option
#'    including "cell", "nature", "science", "lancet", "nejm" and "jama".
#' @return A list object containing the results of normality test for variables and  visualization of the results.
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatDesc(PID = res$PID)
#'    res3 = DescNorm(PID=res$PID, Vars = 'X5,X6,X7,X8,X9', Method = "shapiro.test", Layout = "column" ,
#'    Brightness = "light", Palette = "default2")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

DescNorm = function(PID,
                    OutPath="default",
                    Vars,
                    Method,
                    Layout,
                    Brightness,
                    Palette){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DescNorm')
  res = httr::POST(url,
                   body = list(PID=PID,
                               DescNormVars=Vars,
                               DescNormMethod=Method,
                               DescNormLayout=Layout,
                               DescNormBrightness=Brightness,
                               DescNormPalette=Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/DescNorm'))
  for (name in names(results$normtable)){
    try(
      vroom::vroom_write(results$normtable[[name]],
                         paste0(OutPath,'/DescNorm/', name,".csv"),
                         delim = ","),silent = TRUE
    )
  }

  for (name in names(results$NormPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/DescNorm/',name,".jpg"),
                      plot = results$NormPlot[[name]],
                      width =  results$NormPlot_para[[name]][['width']],
                      height = results$NormPlot_para[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }

  return(results)
}


#' @title Extreme value calculation.
#' @description Extreme value calculation for numeric variables.
#' @usage DescExtre(PID=res$PID, Vars = "X5,X6,X7,X8,X9", LimitLow = 0.025, LimitUpper = 0.975,
#'    Layout = "column.points", Brightness = "light", Palette = "default2").
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param OutPath chr. Output file directory, e.g. "D:/StatDesc/DescExtre". It should be
#'    noted that the slash symbol is "/", not "\".If "default", the current working directory will be set.
#' @param Vars chr. Target variables used for modelling. It should be noted that there is
#'    fixed format for the entering characters separated with  "," and without space.
#'    The default values is "all" (all variables are included).
#' @param LimitLow num. Lower limit ratio to screen the small extreme values located from 0 to this lower limit of the target variables.
#' @param LimitUpper num. Upper limit ratio to screen the large extreme values located from this lower limit to 1 of the target variables.
#' @param Layout chr. Visualization layout . Available values include "column.points", "heatmap".
#' @param Brightness chr. Visualization brightness . Available values include "light" and "dark".
#' @param Palette chr. Visualization palette . Available values include "default1", "default2" and 6 journal option
#'    including "cell", "nature", "science", "lancet", "nejm" and "jama".
#' @return A list object containing the results of extremum for variables and  visualization of the results.
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatDesc(PID = res$PID)
#'    res3 = DescExtre(PID = res$PID, Vars = "X5,X6,X7,X8,X9", LimitLow = 0.025, LimitUpper = 0.975,
#'    Layout = "column.points", Brightness = "light", Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

DescExtre = function(PID,
                     OutPath="default",
                     Vars,
                     LimitLow,
                     LimitUpper,
                     Layout,
                     Brightness,
                     Palette){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DescExtre')
  res = httr::POST(url,
                   body = list(PID=PID,
                               DescExtreVars =Vars,
                               DescExtreLimitLow = LimitLow,
                               DescExtreLimitUpper = LimitUpper,
                               DescExtreLayout =Layout,
                               DescExtreBrightness = Brightness,
                               DescExtrePalette = Palette),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/DescExtre'))
  try(
    vroom::vroom_write(results$Extretable,
                       paste0(OutPath,'/DescExtre/DescExtre.csv'),
                       delim = ","),silent = TRUE
  )

  for (name in names(results$ExtrePlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/DescExtre/',name,".jpg"),
                      plot = results$ExtrePlot[[name]],
                      width =  results$ExtrePlot_para[[name]][['width']],
                      height = results$ExtrePlot_para[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  return(results)
}


#' @title Variable description.
#' @description Calculate median, mean .etc for numeric variables,frequency and count for Discrete variable.
#' @usage DescSize(PID=res$PID, Vars = "C1,C2,X5,X6,X7,X8,X9", VarsBy = "Y1", Layout = "box" , Brightness = "dark" ,
#'    Palette ="science").
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param OutPath chr. Output file directory, e.g. "D:/StatDesc/StatDesc". It should be
#'    noted that the slash symbol is "/", not "\".If "default", the current working directory will be set.
#' @param Vars chr. Target variables used for modelling. It should be noted that there is fixed format for the entering characters
#'    separated with  "," and without space, e.g. "X4,X5,X6,X7,X8". The default values is "all" (all variables are included).
#' @param VarsBy chr. Variable used to group the observation for size description.
#' @param Layout chr. Visualization layout. Available values include "box", "violin".
#' @param Brightness chr. Visualization brightness. Available values include "light" and "dark".
#' @param Palette chr. Visualization palette. Available values include "default1", "default2" and 5 journal option
#'    including "jama", "nature", "science", "lancet", "nejm".
#' @return A list object containing the results of variable description for continuous and discrete variables respectively
#'    and  visualization of the continuous variables.
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatDesc(PID = res$PID)
#'    res3 = DescSize(PID = res$PID, Vars = "X5,X6,X7,X8,X9", VarsBy = "Y1", Layout = "box",
#'    Brightness = "light", Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

DescSize = function(PID,
                    OutPath="default",
                    Vars,
                    VarsBy,
                    Layout,
                    Brightness,
                    Palette){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DescSize')
  res = httr::POST(url,
                   body = list(PID=PID,
                               DescSizeVars=Vars,
                               DescSizeVarsBy=VarsBy,
                               DescSizeLayout=Layout,
                               DescSizeBrightness=Brightness,
                               DescSizePalette=Palette
                   ),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/DescSize'))
  for (name in names(results$Desctable)){
    try(
      vroom::vroom_write(results$Desctable[[name]],
                         paste0(OutPath,'/StatDesc/',name,'.csv'),
                         delim = ","),silent = TRUE
    )
  }

  for (name in names(results$DescPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/DescSize/',name,".jpg"),
                      plot = results$DescPlot[[name]],
                      width =  results$DescPlot_para[[name]][['width']],
                      height = results$DescPlot_para[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  return(results)
}


#' @title Size comparison for groups.
#' @description Perform size comparisons between groups.
#' @usage DescComp(PID = res$PID, Task = "mean", Vars = "X5,X6,X7,X8,X9", VarsBy = "Y1" , Method = "wilcox",
#'    Layout = "density" , Brightness = "dark" , Palette = "default3").
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param OutPath chr. Output file directory, e.g. "D:/StatDesc/DescComp". It should be
#'    noted that the slash symbol is "/", not "\".If "default", the current working directory will be set.
#' @param Task chr. Comparison task. At present, only the "mean" comparison is available.
#' @param Vars chr. Target variables used for modelling. It should be noted that there is fixed format for the entering
#'    characters separated with  "," and without space, e.g. "X5,X6,X7,X8,X9".
#'    The default values is "all" (all variables are included).
#' @param VarsBy chr. Variable  used to group the observation for size comparison.
#' @param Method chr. Comparison method. At present, only "wilcox" (Wilcoxon rank sum test) is available.
#' @param Layout chr. Visualization layout. Available values include "column.points", "density".
#' @param Brightness chr. Visualization brightness. Available values include "light" and "dark".
#' @param Palette chr. Visualization palette. Available values include "default1", "default2" and 6 journal option
#'    including "cell", "nature", "science", "lancet", "nejm" and "jama".
#' @return A list object containing the results of size comparisons between groups for variables and  visualization of the results.
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatDesc(PID = res$PID)
#'    res3 = DescComp(PID=res$PID, OutPath="default",Task = "mean", Vars = "X5,X6,X7,X8,X9", VarsBy = "Y1",
#'    Method = "wilcox", Layout = "column.points",Brightness = "light" , Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

DescComp = function(PID,
                    OutPath="default",
                    Task= "mean",
                    Vars,
                    VarsBy,
                    Method= "wilcox",
                    Layout,
                    Brightness,
                    Palette){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DescComp')
  res = httr::POST(url,
                   body = list(PID=PID,
                               DescCompTask =Task,
                               DescCompVars =Vars,
                               DescCompVarsBy =VarsBy,
                               DescCompMethod =Method,
                               DescCompLayout =Layout,
                               DescCompBrightness =Brightness,
                               DescCompPalette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/DescComp'))
  for (name in names(results$Comptable)){
    try(
      vroom::vroom_write(results$Comptable[[name]],
                         paste0(OutPath,'/DescComp/',name,'.csv'),
                         delim = ","),silent = TRUE
    )
  }

  for (name in names(results$CompPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/DescComp/',name,".jpg"),
                      plot = results$CompPlot[[name]],
                      width =  results$CompPlot_para[[name]][['width']],
                      height = results$CompPlot_para[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  return(results)
}


#' @title Correlation analysis.
#' @description Perform correlation analysis between variables.
#' @usage DescCorr(PID = res$PID, VarsX = "X5,X6,X7,X8,X9", VarsY = "Y1", VarsBy = "Y1", Method = "pearson",
#'    Layout = "bubble", Brightness = "dark", Palette = "nature").
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatDesc.
#' @param OutPath chr. Output file directory, e.g. "D:/StatDesc/DescCorr". It should be
#'    noted that the slash symbol is "/", not "\".If "default", the current working directory will be set.
#' @param VarsX chr. Target variables used for modelling. It should be noted that there is fixed format for the entering
#'    characters separated with "," and without space, e.g. "X5,X6,X7,X8,X9".
#'    The default values is "all.x" (all variables are included).
#' @param VarsY chr. Target outcome variables used for correlation analysis.
#' @param VarsBy chr. Variable  used to group the observation for correlation analysis.
#' @param Method chr.  Method for orrelation analysis. Available values include "spearman" (Spearman's rank correlation analysis)
#'    and "pearson" (Pearson correlation analysis).
#' @param Layout chr. Visualization layout. Available values include "heatmap", "bubble", "matrix".
#' @param Brightness chr. Visualization brightness. Available values include "light" and "dark".
#' @param Palette chr. Visualization palette. Available values include "default1", "default2" , and 6 journal option
#'    including "cell", "nature", "science", "lancet", "nejm" and "jama".
#' @return A list object containing the results of correlation analysis between variables and  visualization of the results.
#' @export
#' @examples res = InitStatDesc()
#'    res1 = LoadStatDesc(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatDesc(PID = res$PID)
#'    res3 = DescCorr(PID = res$PID, VarsX = "X5,X6,X7,X8,X9", VarsY = "Y1", VarsBy = "Y1", Method = "spearman", Layout = "heatmap",
#'    Brightness = "light", Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Yanqiu Feng, Bin Wang (corresponding author)

DescCorr = function(PID,
                    OutPath="default",
                    VarsX,
                    VarsY,
                    VarsBy,
                    Method,
                    Layout,
                    Brightness,
                    Palette){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'DescCorr')
  res = httr::POST(url,
                   body = list(PID=PID,
                               DescCorrVarsX =VarsX,
                               DescCorrVarsY =VarsY,
                               DescCorrVarsBy =VarsBy,
                               DescCorrMethod =Method,
                               DescCorrLayout =Layout,
                               DescCorrBrightness =Brightness,
                               DescCorrPalette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/DescCorr'))
  for (name in names(results$Corrtable)){
    try(
      vroom::vroom_write(results$Corrtable[[name]],
                         paste0(OutPath,'/DescCorr/',name,'.csv'),
                         delim = ","),silent = TRUE
    )
  }

  for (name in names(results$CorrPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/DescCorr/',name,".png"),
                      plot = results$CorrPlot[[name]],
                      width =  results$CorrPlot_para[[name]][['width']],
                      height = results$CorrPlot_para[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  return(results)
}

#' @title Initialize StatCros module
#' @description Initialize StatCros module analysis. It can generate an R6 class object.
#' @usage InitStatCros()
#' @param
#' @details StatCros module was designed to analyze the cross-sectional data from
#'     exposome-wide association study (EWAS). This data structure can be obtained
#'     from the epidemiological designs of cross-section, case-control, and cohort.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatCros()
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

InitStatCros = function(){
  url = paste0(urlhead,'InitCros')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}

#' @title Load data file for StatCros module
#' @description Load data file for StatCros module
#' @usage LoadStatCros(PID, UseExample = "default", DataPath=NULL, VocaPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatCros
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_StatCros.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input directory of vocabulary file, e.g. "D:/test/eg_voca_StatCros.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatCros()
#'    res1 = LoadStatCros(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

LoadStatCros =function(PID,
                       UseExample = "default",
                       DataPath = NULL,
                       VocaPath = NULL){
  url = paste0(urlhead,'LoadCros')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}

  res = httr::POST(url,
                   body = serialize(list(PID = PID,
                                         UseExample = UseExample,
                                         Expodata = exdata,
                                         Vocadata = vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Tidy data.
#' @description Tidy data file for StatCros module
#' @usage TidyStatCros(PID, OutPath = "default",TransDummyVars="default")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatCros
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TransDummyVars chr. Variables to be transformed as dummy variables. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., "X1,X2,X3". If "default", all the factor variables will be transformed into dummy ones.
#'     These variables need to be transformed as factor ones in previous transform step
#'     using TransType function.
#' @return An R6 class object containing the variable(s) after transforming the factor
#'     variables into dummy ones.
#' @export
#' @examples res <- InitStatCros()
#'     res1 = LoadStatCros(PID = res$PID, UseExample = "example#1")
#'     res2 = TidyStatCros(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang
TidyStatCros = function(PID,
                        OutPath = "default",
                        TransDummyVars="default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyCros')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransDummyVars=TransDummyVars),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyStatCros'))
  try(vroom::vroom_write(results$Expo$Data,
                         paste0(OutPath,'/TidyStatCros/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$Expo$Voca,
                         paste0(OutPath,'/TidyStatCros/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Association analysis
#' @description Association analysis for cross-sectional data
#' @usage CrosAsso = function(PID, OutPath = "default",Linear = T, EpiDesign = "cross.sectional",
#'     VarsY, VarsX = "all.cx",VarsN = "single.factor",VarsSel = F,VarsSelThr = 0.1,
#'     Covariates = "all.c",Family,RepMsr = F,Corstr = "ar1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatCros
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Linear lgl. T (or TRUE) and F (or FALSE). Whether the relationship between variables is linear
#' @param EpiDesign chr. Epidemiological design of the study, including "cohort" "case.control"
#'     and "cross.sectional". It doesn’t affect the modeling, but the format of the output file.
#'     For the three designs, the effect values are usually indicated by RR (relative risk) of cohort,
#'     OR (odds ratio) of case-control, and beta value of cross-sectional.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param VarsSel lgl. T (or TRUE) and F (or FALSE). Whether to select the significant variable for the final model.
#'     Available options.
#' @param VarsSelThr num. If "VarsSel" = T, provide the selection threshold
#'     of the P-value. Three value can be chosen, i.e. 0.05, 0.1, and 0.2.
#' @param Covariates chr. Covariates used for modeling. The default option is
#'     "all.c" (All covariates are included).
#' @param Family chr. The link function for the regression model according the data type of outcomes,
#'     including "gaussian" for continuous variable, "binomial" for binary variable,
#'     and "poisson" for counting variable
#' @param RepMsr lgl.  T (or TRUE) and F (or FALSE). Whether existing repeated measurement of the subjects.
#'     Available options.
#' @param Corstr chr. If "RepMsr" = T, the generalized estimating equations (GEE) will be used.
#'     For GEE, three correlation structure options are "exchangeable" "ar1" "unstructured".
#' @details
#' @return A list containing the association analysis results.
#' @export
#' @examples res <- InitStatCros()
#'    res1 = LoadStatCros(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatCros(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = CrosAsso(PID=res$PID, Linear = TRUE, EpiDesign = "cohort",
#'    VarsY = "Y1", VarsX = "X5,X6,X7,X8,X9,X10,X11", VarsN = "single.factor" ,
#'    VarsSel = FALSE, VarsSelThr = 0.1, Covariates = "all.c", Family = "gaussian",
#'    RepMsr = FALSE,Corstr = "ar1")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang
CrosAsso = function(PID,
                    OutPath = "default",
                    Linear = T,
                    EpiDesign = "cross.sectional",
                    VarsY,
                    VarsX = "all.cx",
                    VarsN = "single.factor",
                    VarsSel = F,
                    VarsSelThr = 0.1,
                    Covariates = "all.c",
                    Family,
                    RepMsr = F,
                    Corstr = "ar1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'CrosAsso')
  res = httr::POST(url,
                   body = list(PID = PID,
                               CrosAssoLinear = Linear,
                               CrosAssoEpiDesign = EpiDesign,
                               CrosAssoVarsY = VarsY,
                               CrosAssoVarsX = VarsX,
                               CrosAssoVarsN = VarsN,
                               CrosAssoVarsSel = VarsSel,
                               CrosAssoVarsSelThr = VarsSelThr,
                               CrosAssoCovariates = Covariates,
                               CrosAssoFamily = Family,
                               CrosAssoRepMsr = RepMsr,
                               CrosAssoCorstr = Corstr),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/CrosAsso'))
  for (name in names(results$AssoTable)){
    try(
      vroom::vroom_write(results$AssoTable[[name]],
                         paste0(OutPath,'/CrosAsso/', name,".csv"),
                         delim = ","),
      silent = TRUE
    )
  }
  return(results$AssoTable)
}



#' @title Visualize association analysis
#' @description Visualize association analysis
#' @usage VizCrosAsso = function(PID, OutPath = "default",VarsN = "single.factor",
#'     Layout = "volcano",Brightness = "light",Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatCros
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param Layout chr. Visualization layout. Available options include "forest" and "volcano".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @param ColorFor chr.
#' #' @param SizeFor chr
#' @details
#' @return An R6 class object containing the results' plot.
#' @export
#' @examples res <- InitStatCros()
#'    res1 = LoadStatCros(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatCros(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = CrosAsso(PID=res$PID, Linear = TRUE, EpiDesign = "cohort",
#'    VarsY = "Y1", VarsX = "X5,X6,X7,X8,X9,X10,X11", VarsN = "single.factor" ,
#'    VarsSel = FALSE, VarsSelThr = 0.1, Covariates = "all.c", Family = "gaussian",
#'    RepMsr = FALSE,Corstr = "ar1")
#'    res4 = VizCrosAsso(PID=res$PID, VarsY = "Y1", VarsN="single.factor", Layout = "volcano",
#'    Brightness = "dark",Palette = "default1", ColorFor = "beta.value", SizeFor = "p.value")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

VizCrosAsso = function(PID,
                       OutPath = "default",
                       VarsY,
                       VarsN = "single.factor",
                       Layout = "volcano",
                       Brightness = "light",
                       Palette = "default1",
                       ColorFor = "beta.value",
                       SizeFor = "p.value"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizCrosAsso')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizCrosAssoVarsY = VarsY,
                               VizCrosAssoVarsN = VarsN,
                               VizCrosAssoLayout = Layout,
                               VizCrosAssoBrightness = Brightness,
                               VizCrosAssoPalette = Palette,
                               VizCrosAssoColorFor = ColorFor,
                               VizCrosAssoSizeFor = SizeFor),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (name in names(results$AssoPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/CrosAsso/',name,".png"),
                      plot = results$AssoPlot[[name]],
                      width =  results$AssoPlot_para[[name]][['width']],
                      height = results$AssoPlot_para[[name]][['height']],
                      units = "cm",dpi = 300,
                      limitsize = FALSE),
      silent = TRUE
    )
  }
  return(results$AssoPlot)
}

#' @title Build prediction models
#' @description Build prediction models
#' @usage CrosPred = function(PID,OutPath = "default",VarsY,VarsX = "all.x",
#'     PredType = "response",VarsSel = FALSE,VarsSelThr = 0.1, Covariates = "all.c",
#'     RsmpMethod = "cv",Folds = 5,Ratio = 0.667,Repeats = 5)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatCros
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param PredType chr. Prediction type of the outcome variable, including "response"
#'     for the actual values and "prob" for outcome with binary variable.
#' @param VarsSel lgl. Whether to select the significant variable for the final model.
#'     Available options include T (or TRUE) and F (or FALSE).
#' @param VarsSelThr num. If "VarsSel" = T, provide the selection threshold of
#'     the P-value. Three value can be chosen, i.e. 0.05, 0.1, and 0.2.
#' @param Covariates chr. Covariates used for modeling. The default option is
#'     "all.c" (All covariates are included).
#' @param RsmpMethod chr. Four resampling methods options for internal validation, including
#'     "cv" (i.e.,Cross validation) , "loo" (i.e., eave-one-out), "bootstrap", and "holdout".
#' @param Folds num. Folds of Cross-validation resampling. It is ranging 2-10.
#' @param Ratio num. Ratio of Bootstrap resampling. It is ranging 0.4-0.9.
#' @param Repeats num. Number of Bootstrap resampling. It is ranging 2-20.
#' @details
#' @return A list containing the prediction performance evaluation.
#' @export
#' @examples res <- InitStatCros()
#'    res1 = LoadStatCros(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatCros(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = CrosPred(PID=res$PID, VarsY = "Y1", VarsX = "X5,X6,X7,X8,X9,X10,X11",
#'    PredType = "response",VarsSel = FALSE,VarsSelThr = 0.1,Covariates = "all.c",
#'    RsmpMethod = "cv" ,Folds = 5,Ratio = 0.667,Repeats = 5)
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

CrosPred = function(PID,
                    OutPath = "default",
                    VarsY,
                    VarsX = "all.x",
                    PredType = "response",
                    VarsSel = F,
                    VarsSelThr = 0.1,
                    Covariates = "all.c",
                    RsmpMethod = "cv",
                    Folds = 5,
                    Ratio = 0.667,
                    Repeats = 5){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'CrosPred')
  res = httr::POST(url,
                   body = list(PID = PID,
                               CrosPredVarsY = VarsY,
                               CrosPredVarsX = VarsX,
                               CrosPredPredType = PredType,
                               CrosPredVarsSel = VarsSel,
                               CrosPredVarsSelThr = VarsSelThr,
                               CrosPredCovariates = Covariates,
                               CrosPredRsmpMethod = RsmpMethod,
                               CrosPredFolds = Folds,
                               CrosPredRatio = Ratio,
                               CrosPredRepeats = Repeats),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/CrosPred'))
  for (name in names(results$PredTable)){
    try(
      vroom::vroom_write(results$PredTable[[name]],
                         paste0(OutPath,'/CrosPred/', name,".csv"),
                         delim = ","),
      silent = TRUE
    )
  }
  return(results$PredTable)
}


#' @title Visualize the prediction performance
#' @description Visualize the prediction performance
#' @usage VizCrosPred = function(PID,OutPath = "default",Layout = "bar",
#'     Brightness = "light",Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatCros
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param Layout chr. Visualization layout. Available options include "bar" and "roc"
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing the results' plot.
#' @export
#' @examples res <- InitStatCros()
#'    res1 = LoadStatCros(PID = res$PID, UseExample = "example#1")
#'    res3 = CrosPred(PID=res$PID, VarsY = "Y1", VarsX = "X5,X6,X7,X8,X9,X10,X11",
#'    PredType = "response",VarsSel = FALSE,VarsSelThr = 0.1, Covariates = "all.c",
#'    RsmpMethod = "cv" ,Folds = 5,Ratio = 0.667,Repeats = 5)
#'    res4 = VizCrosPred(PID=res$PID, VarsY = "Y1", Layout = "bar",
#'    Brightness = "light", Palette = "science")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

VizCrosPred = function(PID,
                       OutPath = "default",
                       VarsY,
                       Layout = "bar",
                       Brightness = "light",
                       Palette = "default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  url = paste0(urlhead,'VizCrosPred')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VizCrosPredVarsY = VarsY,
                               VizCrosPredLayout = Layout,
                               VizCrosPredBrightness = Brightness,
                               VizCrosPredPalette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (name in names(results$PredPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/CrosPred/', name, ".png"),
                      plot = results$PredPlot[[name]],
                      width = 25,
                      height = 14,
                      units = "cm",
                      dpi = 300),
      silent = TRUE
    )
  }

  return(results$PredPlot)
}


#' @title Initialize StatPanel module
#' @description Initialize StatPanel module analysis. It can generate an R6 class object.
#' @usage InitStatPanel()
#' @param
#' @details StatPanel module is designed to conduct the analysis of the panel data.
#'     It mainly aims to evaluate the associations between exposure factors and the health outcome.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatPanel()
#'     FuncExit(PID = res$PID)
#' @author Bin Wang (corresponding author)

InitStatPanel = function(){
  url = paste0(urlhead,'InitPanel')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for StatPanel module
#' @description Load data file for StatPanel module
#' @usage LoadStatPanel(PID, UseExample = "default", DataPath=NULL, VocaPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatPanel
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input directory of vocabulary file, e.g. "D:/test/eg_voca_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatPanel()
#'    res = LoadStatPanel(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

LoadStatPanel =function(PID,
                        UseExample = "default",
                        DataPath = NULL,
                        VocaPath = NULL){
  url = paste0(urlhead,'LoadPanel')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)

    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID = PID,
                                         UseExample = UseExample,
                                         Expodata = exdata,
                                         Vocadata = vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Transform factor variables into dummy ones
#' @description Transform factor variables into dummy ones
#' @usage TidySataPanel(PID, OutPath = "default",TransDummyVars="default")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatPanel
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TransDummyVars chr. Variables to be transformed as dummy variables. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., "X1,X2,X3". If "default", all the factor variables will be transformed into dummy ones.
#'     These variables need to be transformed as factor ones in previous transform step
#'     using TransType function.
#' @return An R6 class object containing the variable(s) after transforming the factor
#'     variables into dummy ones.
#' @export
#' @examples res <- InitStatPanel()
#'     res1 = LoadStatPanel(PID = res$PID, UseExample = "example#1")
#'     res2 = TidySataPanel(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang
TidySataPanel = function(PID,
                         OutPath = "default",
                         TransDummyVars="default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyPanel')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransDummyVars=TransDummyVars),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidySataPanel'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidySataPanel/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidySataPanel/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Association analysis of panel data
#' @description Association analysis of panel data
#' @usage PanelAsso = function(PID, OutPath = "default", VarsY, VarsX, VarsN = "single.factor",
#'     VarsRandomIpt = "SubjectID", VarsRandomSlp = "none",Covariates = "all.c")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatPanel
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param VarsRandomIpt chr. Random intercept variable for the linear mixed-effect model.
#'     The default is "SubjectID".
#' @param VarsRandomSlp chr. Random slope variable for the linear mixed-effect model.
#'     The default is "none". It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param Covariates chr. Covariates used for modeling. The default option is
#'     "all.c" (All covariates are included).
#' @details
#' @return A list containing the association analysis results.
#' @export
#' @examples res <- InitStatPanel()
#'     res1 = LoadStatPanel(PID = res$PID, UseExample = "example#1")
#'     res2 = TidySataPanel(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     res3 = PanelAsso(PID=res$PID, OutPath = "default", VarsY = "Y1",
#'     VarsX = "X1,X2,X3,X4,X5,X6,X7,X8,X9,X10,X11,X12", VarsN = "single.factor",
#'     VarsRandomIpt = "SubjectID", VarsRandomSlp = "none", Covariates = "all.c")
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

PanelAsso = function(PID,
                     OutPath = "default",
                     VarsY,
                     VarsX,
                     VarsN = "single.factor",
                     VarsRandomIpt = "SubjectID",
                     VarsRandomSlp = "none",
                     Covariates = "all.c"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'PanelAsso')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsX = VarsX,
                               VarsN = VarsN,
                               VarsRandomIpt = VarsRandomIpt,
                               VarsRandomSlp = VarsRandomSlp,
                               Covariates = Covariates),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/PanelAsso'))
  for (name in names(results$AssoTable)){
    try(
      vroom::vroom_write(results$AssoTable[[name]],
                         paste0(OutPath,'/PanelAsso/', name,".csv"),
                         delim = ","),
      silent = TRUE
    )
  }
  return(results$AssoTable)
}


#' @title Visualize the results of association analysis for panel data
#' @description Visualize the results of association analysis for panel data
#' @usage VizPanelAsso = function(PID, OutPath, VarsY, VarsN, Layout = "volcano",
#'     Brightness = "dark", Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatPanel
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param Layout chr. Visualization layout. Available options include "forest" and "volcano".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing the plots of association analysis results.
#' @export
#' @examples res <- InitStatPanel()
#'     res1 = LoadStatPanel(PID = res$PID, UseExample = "example#1")
#'     res2 = TidySataPanel(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     res3 = PanelAsso(PID=res$PID, OutPath = "default", VarsY = "Y1",
#'     VarsX = "X1,X2,X3,X4,X5,X6,X7,X8,X9,X10,X11,X12", VarsN = "single.factor",
#'     VarsRandomIpt = "SubjectID", VarsRandomSlp = "none", Covariates = "all.c")
#'     res4 = VizPanelAsso(PID = res$PID, OutPath = "default", VarsY = "Y1",
#'     VarsN = "single.factor", Layout = "forest", Brightness = "dark",Palette = "default1")
#'     FuncExit(PID = res$PID)
#' @author Bin Wang

VizPanelAsso = function(PID,
                        OutPath = "default",
                        VarsY,
                        VarsN,
                        EffectThr = 0.5,
                        Layout = "volcano",
                        Brightness = "dark",
                        Palette = "default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'VizPanelAsso')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsN = VarsN,
                               EffectThr = EffectThr,
                               Layout = Layout,
                               Brightness = Brightness,
                               Palette = Palette),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (name in names(results$AssoPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/PanelAsso/',name,".png"),
                      plot = results$AssoPlot[[name]],
                      width = results$AssoPlot_para[[name]][['width']],
                      height = results$AssoPlot_para[[name]][['height']],
                      units = "cm",dpi = 300,
                      limitsize = FALSE),
      silent = TRUE
    )
  }
  return(results$AssoPlot)
}



#' @title Initialize EVIZ module
#' @description Initialize EVIZ module analysis. It can generate an R6 class object
#'     integrating all the analysis information.
#' @usage InitEVIZ()
#' @param
#' @details EVIZ module is designed for the data visualization of different
#'     statistical and biological analyses in a user friendly and easy way, including
#'     four typical classes of visualization.
#' @return An R6 class object.
#' @export
#' @examples res <- InitEVIZ()
#' @author Ning Gao, Bin Wang (corresponding author)

InitEVIZ = function(){
  url = paste0(urlhead,'InitViz')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for EVIZ module
#' @description Load data for visualization.
#' @usage LoadEVIZ(PID, UseExample = "default", FileDirIn=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input data file directory, e.g. "D:/test/eg_expoviz_data.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input vocabulary file directory, e.g. "D:/test/eg_expoviz_voca.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#' @author Ning Gao, Bin Wang (corresponding author)

LoadEVIZ =function(PID,
                   UseExample="default",
                   DataPath=NULL,
                   VocaPath=NULL){
  url = paste0(urlhead,'LoadViz')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata,
                                         Vocadata=vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}

#' @title Plot category dot
#' @description Visualize data via dot plot.
#' @usage EVIZCateDot(PID,OutPath,Vars, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Vars chr. Specifying the variables.
#'     Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#'     No more than 50 variables be entered is recommended (< 50 variables).
#' @param Parameter chr. Specifying which parameter of the data to be the ordinate of
#'     the output plot. Available options include: "mean", "median", "min", "max", "mad"
#'     or "sd".Default is "mean".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details The dot plot is used to display the relative position of
#'     two data points in the same time period, or compare the difference
#'     between the two categorical variables.
#' @return A list containing the dot plot.
#'     "light_default1": the visualization result.
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2 = EVIZCateDot(PID=res$PID,Vars="X4,X5,X6,X7,X8,X9,X10",Parameter="mean",Brightness="light",Palette="default1")
#' @author Ning Gao,Bin Wang(corresponding author)

EVIZCateDot = function(PID,
                       OutPath="default",
                       Vars,
                       Parameter="mean",
                       Brightness="light",
                       Palette='default1'){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizCate_Dot')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizCate_DotVars=Vars,
                               VizCate_DotParameter=Parameter,
                               VizCate_DotBrightness=Brightness,
                               VizCate_DotPalette=Palette),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Cate'))
  for (name in names(results$Cate_Dot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Cate/',Parameter,'_',name,".png"),
                      plot = results$Cate_Dot[[name]] ,
                      width =  results$Dot_plotpara$PlotWidth,
                      height = results$Dot_plotpara$PlotHeight,
                      units = "cm"),silent = TRUE
    )
  }
  return(results$Cate_Dot)
}


#' @title Plot relationship network
#' @description Visualize data via network plot.
#' @usage EVIZRelatNetwork(PID,OutPath,VarsY,VarsX, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Specifying the dependent variables(e.g."Y2").
#' @param VarsX chr. Specifying the independent variables.Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#' @param Family chr. Specifying the data distribution in order to determine the visualization method.
#'      Available options include:"gaussian","poisson" and "logistic".Notice that the family are
#'     determined by data type of an outcome, or the plot can not be visualized.
#' @param Layout chr. Visualization layout. Available options include "force-directed" and "degree-circle".
#' @param CutOff num. Partial outcomes to visualize which is determined by correlation
#'     coefficient r. The range must between 0 and 1.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details The network plot is used to visualize the relationship between
#'     the input variables by using ggraph package.
#' @return A list containing one plot.
#'     (1) "light_default1": the visualization result;
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2 = EVIZRelatNetwork(PID=res$PID,VarsY="Y1",VarsX="all.x",Family="gaussian",Layout="force-directed",CutOff=0.8,Brightness="light",Palette="default1")
#' @author Ning Gao,Bin Wang(corresponding author)

EVIZRelatNetwork = function(PID,
                            OutPath="default",
                            VarsY,
                            VarsX="all.x",
                            Family="gaussian",
                            Layout="force-directed",
                            CutOff=0.8,
                            Brightness="light",
                            Palette="default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra)
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'VizRelat_Network')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizRelat_NetworkVarsY=VarsY,
                               VizRelat_NetworkVarsX=VarsX,
                               VizRelat_NetworkFamily=Family,
                               VizRelat_NetworkLayout=Layout,
                               VizRelat_NetworkCutOff=CutOff,
                               VizRelat_NetworkBrightness=Brightness,
                               VizRelat_NetworkPalette=Palette),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Network'))
  for (name in names(results$Relat_Network)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Network', "/network_",Layout,"_",
                             Brightness,"_",
                             Palette,".png"),
                      plot = results$Relat_Network[[name]],
                      width =  20,
                      height = 20,
                      units = "cm"),silent = TRUE
    )
  }
  return(results$Relat_Network)
}


#' @title Plot elationship edge bundling
#' @description Visualize data via edge bundling plot.
#' @usage EVIZRelatEdgeBundling(PID,OutPath,VarsY,VarsX, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Dependent variables for visualization(e.g."Y2").
#' @param VarsX chr. Independent variables.Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#' @param Family chr. Specifying the data distribution in order to determine the visualization method.
#'      Available options include:"gaussian","poisson" and "logistic".Notice that the family are
#'     determined by data type of an outcome, or the plot can not be visualized.
#' @param SizeFor chr. Parameter to represent the size of the points in the output plot. Available options
#'     include "pvalue" and "beta".The default option is "pvalue".
#' @param CutOff num. Partial outcomes to  visualize which is determined by correlation coefficient r.
#'     The range must between 0 and 1.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details The edge bundling plot is used to bundle the edges closely
#'     in order to reduce complexity.
#' @return A list containing one plot.
#'     (1) "light_default1": the visualization result;
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2 = EVIZRelatEdgeBundling(PID=res$PID,VarsY = "Y2",VarsX = "all.x",Family = "gaussian" ,SizeFor = "pvalue",Brightness = "light",Palette = "default1")
#' @author Ning Gao,Bin Wang(corresponding author)

EVIZRelatEdgeBundling = function(PID,
                                 OutPath="defalut",
                                 VarsY,
                                 VarsX="all.x",
                                 Family="gaussian",
                                 SizeFor="pvalue",
                                 Brightness="light",
                                 Palette="default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra)
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'VizRelat_EdgeBundling')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizRelat_EdgeBundlingVarsY=VarsY,
                               VizRelat_EdgeBundlingVarsX = VarsX,
                               VizRelat_EdgeBundlingFamily = Family,
                               VizRelat_EdgeBundlingSizeFor = SizeFor,
                               VizRelat_EdgeBundlingBrightness = Brightness,
                               VizRelat_EdgeBundlingPalette =Palette),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/EdgeBundling'))
  for (name in names(results$Relat_EdgeBundling)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/EdgeBundling', "/EdgeBundling_",SizeFor,
                             "_",Brightness,"_",Palette,".png"),
                      plot = results$Relat_EdgeBundling[[name]],
                      width =  20,
                      height = 20,
                      units = "cm"),silent = TRUE
    )
  }
  return(results$Relat_EdgeBundling)
}


#' @title Plot relationship heatmep
#' @description Visualize data via heatmap plot.
#' @usage EVIZRelatHeatmap(PID,OutPath,Vars, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Vars chr. Independent variables.Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#' @param Method chr. Method to calculate the correlation. Default option is "spearman".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details The heatmap plot is used to display data in color changes
#'     as a matrix.
#' @return A list containing one plot.
#'     (1) "light_default1": the whole dataset visualization result.
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2 = EVIZRelatHeatmap(PID=res$PID,Vars = "X1,X4,X5,X6,X7,X8,X9,X10",Method = "spearman",Brightness = "light",Palette = "default1")
#' @author Ning Gao, Bin Wang (corresponding author)

EVIZRelatHeatmap = function(PID,
                            OutPath="default",
                            Vars,
                            Method="spearman",
                            Brightness="light",
                            Palette="default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra)
  url = paste0(urlhead,'VizRelat_Heatmap')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizRelat_HeatmapVars=Vars,
                               VizRelat_HeatmapMethod=Method,
                               VizRelat_HeatmapBrightness=Brightness,
                               VizRelat_HeatmapPalette=Palette),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Heatmap'))
  for (i in c(1:length(results$Relat_Heatmap))){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Heatmap', "/Heatmap_",names(results$Relat_Heatmap)[i],"_",Method,".png"),
                      results$Relat_Heatmap[[i]] ,
                      width  =  results$Heatmap_para[[i*2-1]],
                      height  =  results$Heatmap_para[[i*2]],
                      limitsize = FALSE,
                      units = "cm"),silent = TRUE
    )
  }
  return(results$Relat_Heatmap)
}


#' @title Plot relationship matrix
#' @description Visualize data via matrix plot.
#' @usage EVIZRelatMatrix(PID,OutPath,Vars, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Vars chr. Independent variables.Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#' @param Method chr. Method to calculate the correlation. Default option is "spearman".
#' @details The matrix plot is used to make a matrix of plots with a
#'     given data set.
#' @return A list containing one plot.
#'     (1) "light_default1": the whole dataset visualization result.
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2 = EVIZRelatMatrix(PID=res$PID,Vars = "X4,X5,X6,X7,X8,X9,X10",Method = "spearman")
#' @author Ning Gao, Bin Wang (corresponding author)

EVIZRelatMatrix = function(PID,
                           OutPath="default",
                           Vars,
                           Method="spearman"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra)
  url = paste0(urlhead,'VizRelat_Matrix')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizRelat_MatrixVars =Vars,
                               VizRelat_MatrixMethod = Method
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Matrix'))
  try(
    ddpcr::quiet(ggplot2::ggsave(paste0(OutPath, "/Matrix/Matrix_",Method,".png"),
                                 results$Relat_Matrix,
                                 width  =  results$Relat_Matrix_para[["PlotWidth"]],
                                 height  =  results$Relat_Matrix_para[["PlotHeight"]],
                                 units = "cm")),silent = TRUE

  )
  return(results$Relat_Matrix)
}


#' @title Plot distribution sierra
#' @description Visualize data via sierra plot.
#' @usage EVIZDistrSierra(PID,OutPath,Vars, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Vars chr. Variables to be visualized(e.g."X4,X5,X6,X7,X8,X9,X10").Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#'     No more than 50 variables be entered is recommended (< 20 variables).
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details The sierra plot is used to visualize the kernel density
#'     estimation of data.
#' @return A list containing one plot.
#'     (1) "light_default1": the whole dataset visualization result.
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2 = EVIZDistrSierra(PID=res$PID,Vars = "X14,X15,X16,X17,X18,X19,X20",Brightness = "light",Palette = "default1")
#' @author Ning Gao,Bin Wang(corresponding author)

EVIZDistrSierra = function(PID,
                           OutPath="default",
                           Vars,
                           Brightness="light",
                           Palette="default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizDistr_Sierra')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizDistr_SierraVars=Vars,
                               VizDistr_SierraBrightness=Brightness,
                               VizDistr_SierraPalette=Palette),
                   encode = 'json')

  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/DistrSierra'))
  for (name in names(results$Sierra)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/DistrSierra/','Sierra_',name,".png"),
                      plot = results$Sierra[[name]] ,
                      width =  results$Sierra_para$PlotWidth,
                      height = results$Sierra_para$PlotHeight,
                      units = "cm"),silent = TRUE
    )
  }
  return(results$Sierra)
}


#' @title Plot component dendrogram
#' @description Visualize data via dendrogram plot.
#' @usage EVIZCompoDendrogram(Vars, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEVIZ.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Vars chr. Variables to be visualized(e.g."X4,X5,X6,X7,X8,X9,X10").Available options include:
#'     "all.x", all independent variables;
#'     or input a character string specifying the variables,separated by comma "," without space(e.g."X4,X5,X6,X7,X8,X9,X10").
#' @param Parameter chr. Specifying which parameter of the data to be the ordinate of
#'     the output plot. Available options include: "mean", "median", "min", "max", "mad"
#'     or "sd".Default is "mean".
#' @param DistMethod chr.The distance measure. This must be one of "euclidean", "maximum"
#'     or "manhattan".Default is "euclidean".
#' @param ClusterMethod chr.The agglomeration method. This should be one of "ward.D",
#'     "ward.D2" or "single".Default is "ward.D".
#' @param ClusterNum num. The number of groups for cutting the tree.Default is 4.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details The dendrogram plot is used to plot beautiful dendrograms.
#' @return A list containing one plot.
#'     (1) "light_default1": the whole dataset visualization result.
#' @export
#' @examples res = InitEVIZ()
#'     res1 = LoadEVIZ(PID = res$PID, UseExample = "example#1")
#'     res2= EVIZCompoDendrogram(PID=res$PID,Vars = "X4,X5,X6,X7,X8,X61,X66,X67,X200",Parameter = "median",DistMethod = "euclidean",ClusterMethod = "ward.D2",ClusterNum = "4",Brightness = "light",Palette = "default1")
#' @author Ning Gao,Bin Wang(corresponding author)

EVIZCompoDendrogram = function(PID,
                               OutPath="default",
                               Vars,
                               Parameter="median",
                               DistMethod="euclidean",
                               ClusterMethod="ward.D",
                               ClusterNum=4,
                               Brightness="light",
                               Palette="default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizCompo_Dendrogram')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizCompo_DendrogramVars=Vars,
                               VizCompo_DendrogramParameter=Parameter,
                               VizCompo_DendrogramDistMethod=DistMethod,
                               VizCompo_DendrogramClusterMethod=ClusterMethod,
                               VizCompo_DendrogramClusterNum=ClusterNum,
                               VizCompo_DendrogramBrightness=Brightness,
                               VizCompo_DendrogramPalette=Palette),
                   encode = 'json')

  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Dendrogram'))
  for (name in names(results$Dendrogram)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Dendrogram/Dendrogram_',name,'_',
                             Parameter,'_',
                             DistMethod,'_',
                             ClusterMethod,'_',name,".png"),
                      plot = results$Dendrogram[[name]] ,
                      width =  results$Dendrogram_para$PlotWidth,
                      height = results$Dendrogram_para$PlotHeight,
                      units = "cm"),silent = TRUE
    )
  }
  return(results$Dendrogram)
}

#' @title Initialize StatIP module
#' @description Initialize StatIP module analysis. It can generate an R6 class object.
#' @usage InitStatIP()
#' @param
#' @details StatIP module is designed to find the statistical relationships
#'     between exposure factors and health outcome.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatIP()
#'   FuncExit(PID = res$PID)
#' @author Bin Wang

InitStatIP = function(){
  url = paste0(urlhead,'InitStatIP')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for StatIP for module
#' @description Load data file for StatIP module
#' @usage LoadStatIP(PID, UseExample = "default", DataPath=NULL, VocaPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatIP
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input directory of vocabulary file, e.g. "D:/test/eg_voca_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatIP()
#'    res = LoadStatIP(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

LoadStatIP =function(PID,
                     UseExample="default",
                     DataPath=NULL,
                     VocaPath=NULL){
  url = paste0(urlhead,'LoadStatIP')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)

    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata,
                                         Vocadata=vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}

#' @title
#' @description
#' @usage TidyStatIP(PID, OutPath = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @details
#' @return An R6 class object containing the variable(s) with acceptable variance.
#' @export
#' @examples res = InitStatIP()
#'     res1 = LoadStatIP(PID=res$PID, UseExample="example#1")
#'     res1$Expo$Data
#'     res2 = TidyStatIP(PID=res$PID,OutPath = "default")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang
TidyStatIP = function(PID,
                      OutPath = "default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyStatIP')
  res =  httr::POST(url,
                    body = list(PID=PID),
                    encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyStatIP'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyStatIP/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyStatIP/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}

#' @title Build statistical link for cross-sectional data.
#' @description Build statistical link for cross-sectional data.
#' @usage StatIPCros(PID, OutPath = "default", VarsY, VarsX, LinkModel = "ranger",
#'     ObsrPartType = "raw",ObsrPartNum = "50",ObsrProfType = "partial",
#'     ObsrProfNum = "100",ObsrProfVars = "all.x",ObsrProfGeom = "profiles",
#'     SubjPredSeq = "none",SubjPartType = "break_down")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatIP
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modelling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modelling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param LinkModel chr. Methods to interpret the model. Options include "ranger" (random forest),
#'     "glmnet" (elastic net), "svm" (support vector machine), "glm" (linear regression),
#'     "gam" (generalized additive model), and "xgboost" (eXtreme gradient boosting).
#' @param ObsrPartType chr. Type of transformation that should be applied for dropout loss.
#'     Options include "raw" (drop losses), "ratio" (drop_loss/drop_loss_full_model),
#'     and "difference" (drop_loss - drop_loss_full_model)
#' @param ObsrPartNum chr. Number of observations that should be sampled for calculation
#'     of variable importance. The default means variable importance will be calculated
#'     on whole dataset (no sampling). If "defult", use all Obsrrvations.
#' @param ObsrProfType chr. Type of variable profile. Options include "partial",
#'     "conditional", and "accumulated"
#' @param ObsrProfNum int. Number of observations used for calculation of aggregated profiles.
#'     By default 100.
#' @param ObsrProfVars chr. Names of variables to be explained. If "all.x", all "X variable" above are chosen.
#' @param ObsrProfGeom chr. Layout of the explanation profile in dataset level
#'     including "aggregates", "profiles" or "points".
#' @param SubjPredSeq chr. Subjects which need explanation. Options include "all" (all the subjects),
#'     "none" (no subjects), and "other" (copy the subject list by clicking "Available vars").
#' @param SubjPartType chr. Layout of the explanation profile in subject level.
#'     Options include "shap", "oscillations", "break_down", and "none".
#' @details
#' @return A list containing all the statistical explanation results.
#' @export
#' @examples res <- InitStatIP()
#'    res1 = LoadStatIP(PID = res$PID, UseExample = "example#1")
#'    res2 = StatIPCros(PID=res$PID, OutPath = "default", VarsY = "Y1" ,VarsX = "all.x",
#'    LinkModel = "ranger",ObsrPartType = "raw" ,ObsrPartNum  = "50",
#'    ObsrProfType = "partial" ,ObsrProfNum = "100", ObsrProfVars = "all.x",
#'    ObsrProfGeom = "profiles",SubjPredSeq = "S1,S2,S3",SubjPartType = "break_down")
#'    FuncExit(PID = res$PID)
#' @author Bin Wang

StatIPCros = function(PID,
                      OutPath = "default",
                      VarsY,
                      VarsX,
                      LinkModel = "ranger",
                      ObsrPartType = "raw",
                      ObsrPartNum = "50",
                      ObsrProfType = "partial",
                      ObsrProfNum = "100",
                      ObsrProfVars = "all.x",
                      ObsrProfGeom = "profiles",
                      SubjPredSeq = "none",
                      SubjPartType = "break_down"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'StatIPCros')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsX = VarsX,
                               LinkModel = LinkModel,
                               ObsrPartType = ObsrPartType,
                               ObsrPartNum = ObsrPartNum,
                               ObsrProfType = ObsrProfType,
                               ObsrProfNum = ObsrProfNum,
                               ObsrProfVars = ObsrProfVars,
                               ObsrProfGeom = ObsrProfGeom,
                               SubjPredSeq = SubjPredSeq,
                               SubjPartType = SubjPartType),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/Dataset_Level'))
  for (name in names(results$CrosVipTable)){
    try(
      vroom::vroom_write(results$CrosVipTable[[name]],
                         paste0(OutPath,'/Dataset_Level/', name,".csv"),
                         delim = ","),silent = TRUE
    )
  }

  for (name in names(results$CrosVipPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Dataset_Level/',name,".png"),
                      plot = results$CrosVipPlot[[name]],
                      width = 35,
                      height = 20,
                      units = "cm",dpi = 300),silent = TRUE
    )
  }

  for (name in names(results$CrosProfPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Dataset_Level/',name,".png"),
                      plot = results$CrosProfPlot[[name]],
                      width =  35,
                      height = 20,
                      units = "cm",dpi = 300),silent = TRUE
    )
  }

  for (name in names(results$CrosProfPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Dataset_Level/',name,".png"),
                      plot = results$CrosProfPlot[[name]],
                      width =  35,
                      height = 20,
                      units = "cm",dpi = 300),silent = TRUE
    )
  }

  dir.create(paste0(OutPath,'/Subject_Level'))
  for (name in names(results$CrosSubjPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/Subject_Level/',name,".png"),
                      plot = results$CrosSubjPlot[[name]],
                      width =  35,
                      height = 20,
                      units = "cm",dpi = 300),silent = TRUE
    )
  }

  return(results)
}

#' @title Initialize StatMedt module
#' @description Initialize StatMedt module analysis. It can generate an R6
#'     class object integrating all the analysis information.
#' @usage InitStatMedt()
#' @details InitStatMedt uses R6 package to generate an R6 class object where
#'    parameters to be used for the following mediation module program are
#'    initialized and save in that object. All executed function codes in the
#'    ExpoMediaiton packaged will be recorded in the form of log text in that
#'    object. Furthermore, a program ID (i.e., PID) is randomly created for the
#'    users to identify their own program.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatMedt()
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)
InitStatMedt = function(){
  url = paste0(urlhead,'InitMedt')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for Mediation module
#' @description Load data file for Mediation module
#' @usage LoadStatMedt(PID = PID, UseExample = "default", ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by
#'    InitStatMedt.
#' @param UseExample chr. Method of uploading data. If "default",
#'    user should upload their own data files, or use "example#1"
#'    provided by this module.
#' @param DataPath chr. Input data file directory, e.g.
#'    "D:/test/eg_Medt_data.xlsx". It should be noted that the slash
#'    symbol is "/", not "\".
#' @param VocaPath chr. Input vocabulary file directory, e.g.
#'    "D:/test/eg_Medt_data.xlsx". It should be noted that the slash
#'    symbol is "/", not "\".
#' @details LoadStatMedt function loads the data file and the vocabulary file into
#'    the R6 object that InitStatMedt created. Noted that there are several data
#'    format requirments for the data and vocabulary file.
#'    For data file, the first three column must be named as "SampleID", "SubjectID"
#'    and "Group" in sequence. The "Gourp" variable should be a character variable
#'    to category data into two groups: "train" and "test" group. Outcome variables
#'    should be named as "Y*". e.g., Y1, Y2, Y3... Similarly, exposure variables
#'    should be named as "X*". e.g., X1, X2, X3..., and mediator variables should
#'    be named as "M*". e.g., M1, M2, M3...
#'    For vocabulary file, the first column should be a character variable named
#'    "SerialNo" indicating the names of outcome, exposure and mediator variables
#'    in the data file. These names should be consistent with the variable names
#'    in data file. The second column should be a character variable named "FullName"
#'    indicating the full names (labels) of the variables. The third column should
#'    be a character variable named "SubgroupName" indicating the groups the exposure
#'    or mediator variables belong to.
#' @return An R6 class object containing the input data and vocabulary file.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

LoadStatMedt =function(PID,
                       UseExample="default",
                       DataPath=NULL,
                       VocaPath=NULL){
  url = paste0(urlhead,'LoadMedt')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    # exdata_json = jsonlite::toJSON(exdata)
    # exdata_json = as.character(exdata_json)


    vodata = readxl::read_excel(VocaPath)
    # vodata_json = jsonlite::toJSON(vodata,dataframe = 'columns')
    # vodata_json = as.character(vodata_json)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata,
                                         Vocadata=vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}

#' @Title Tidy data for Mediation module.
#' @description Tidy data for Mediation module.
#' @usage TidyStatMedt(PID, OutPath = "default", LogTransM = LogTransM, RangeLow = RangeLow, RangeUpp = RangeUpp)
#' @param PID chr. Program ID. It must be the same with the PID generated by ExpoMedt
#' @param LogTransM chr. Methods for performing log-transformation. Options include
#'     "log2", "log10", and "ln".
#' @param RangeLow num. Lower range for features to be scaled. Noted that the
#'     value of RangeLow should be lower than that of RangeUpp.
#' @param RangeUpp num. Upper range for features to be scaled. Noted that the
#'     value of RangeUpp should be lower than that of RangeLow.
#' @return
#' @export
#' @examples res <- InitStatMedt()
#'     res1 = LoadStatMedt(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)
TidyStatMedt = function(PID,
                        OutPath = "default",
                        LogTransM = LogTransM,
                        RangeLow = RangeLow,
                        RangeUpp = RangeUpp){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyMedt')
  res = httr::POST(url,
                   body = list(PID=PID,
                               LogTransM = LogTransM,
                               RangeLow = RangeLow,
                               RangeUpp = RangeUpp),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyStatMedt'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyStatMedt/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyStatMedt/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}

#' @title Implement pairwise mediation analyses
#' @description Implement pairwise mediation analyses for each pair of exposure
#'    and mediator.
#' @usage Pairwise(PID, OutPath, VarsY, Family, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default", the
#'    current working directory will be set.It should be noted that the slash
#'    symbol is "/", not "\".
#' @param VarsY chr. The outcome variable. Either continuous, binary or count
#'    type is permitted.
#' @param VarsX chr. Exposure variables included in pairwise mediation modelling.
#'    When "default" is specified, all exposure variables in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "X1,X2,X3".
#' @param VarsM chr. Mediator variables included in pairwise mediation modelling.
#'    When "default" is specified, all mediator variables in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "M1,M2,M3".
#' @param VarsC chr. Covariates included in pairwise mediation modelling.
#'    When "default" is specified, all covariates in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "C1,C2".
#' @param Family chr. The error distribution and link function to be used in the
#'    model. Available options include "linear" and "gaussian" for continuous
#'    outcome variables, and "binomial" as well as "poisson" options for binary
#'    and count outcome variables.
#' @param Iter num. The number of iteration times. Default is 500. To get a more stable
#'    result, a minimum iteration of 1,000 is recommended. To obtain more digits
#'    of P value of parameter estimation, a minimum iteration of 5,000 is suggested.
#' @details Pairwise function implements mediation modelling for each pair of exposure
#'    and mediator. Given M exposures and N mediators, an exhaustive rule will be executed
#'    and M*N pair-wised mediation modelling are fitted. The modelling was realized using
#'    mediate function in mediation package.
#' @return A list containing one dataframe that contains the pairwise mediation modelling results.
#'    1. "MedtPairWise_Stats": pairwise mediation modelling results.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res3 <- Pairwise(PID=res$PID, VarsY = "Y1",
#'    VarsX = "X1,X2,X3", VarsM = "M1,M2,M3", VarsC = "C1", Family = "linear",
#'    Iter = 500)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

Pairwise = function(PID,
                    OutPath = "default",
                    VarsY = "Y1",
                    VarsX = "default",
                    VarsM = "default",
                    VarsC = "default",
                    Family = "linear",
                    Iter = 500){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'Pairwise')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MedtPairWiseVarsY=VarsY,
                               MedtPairWiseVarsX=VarsX,
                               MedtPairWiseVarsM=VarsM,
                               MedtPairWiseVarsC=VarsC,
                               MedtPairWiseFamily=Family,
                               MedtPairWiseIter=Iter
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/1-Pairwise")
  if(!file.exists(path)){dir.create(path)}
  vroom::vroom_write(results$MedtPairWise_Stats,paste0(path, "/Pairwise.modelling.results.csv"),",")
  return(results)
}


#' @title Visualize pairwise mediation modelling result
#' @description Visualize the pairwise mediation result. It should be noted that the
#'    pairwise mediation modelling result has been built by Pairwise function
#'    prior to using it.
#' @usage VizMedtPair(PID, OutPath, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default",
#'    the current working directory will be set.It should be noted that the
#'    slash symbol is "/", not "\".
#' @param Brightness chr. Visualization brightness. Available options include
#'    "light" and "dark".
#' @param Pallette chr. Visualization palette. Available options include "default1",
#'    "default2" and "Journal". The "Journal" option provides several journal
#'    preference styles including cell, nature, science, lancet, nejm, and jama.
#' @details VizMedtPair draws a visualized display for pairwise modelling results.
#'    Piror to using VizMedtPair, make sure that the users have built mediation
#'    models by Pairwise function.
#' @return A list containing two elements where the visualized pairwise modelling plot
#'    as well as the organized plotting data are stored. The elements of that list
#'    include:
#'    1. "plotdata": a dataframe containing the organized plotting data extracted from #'     pairwise mediation modelling result.
#'    2. "plot": a visualized plot for pairwise mediation result.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res3 <- Pairwise(PID=res$PID, VarsY = "Y1",
#'    VarsX = "X1,X2,X3", VarsM = "M1,M2,M3", VarsC = "C1", Family = "linear",
#'    Iter = 500)
#'    res4 <- VizMedtPair(PID=res$PID, Brightness = "light",
#'    Pallette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

VizMedtPair = function(PID,
                       OutPath = "default",
                       Brightness = "light",
                       Pallette = "default1"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizMedtPair')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizMedtPairBrightness=Brightness,
                               VizMedtPairPallette=Pallette),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/1-Pairwise")
  if(!file.exists(path)){dir.create(path)}
  vroom::vroom_write(results$plotdata,paste0(path, "/VizPairwise.plotdata.csv"),",")

  jpeg(paste0(path, "/PairWise_plot.jpg"),
       width = 2000,
       height = 1200,
       res = 220)
  grid::grid.draw(results$plot[[1]])
  dev.off()

  return(results)
}


#' @title Expoures dimension reduction
#' @description Implement dimension reduction for exposures.
#' @usage RedX(PID, OutPath, VarsY, Family, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default",
#'    the current working directory will be set.It should be noted that the
#'    slash symbol is "/", not "\".
#' @param VarsY chr. The outcome variable. Either continuous, binary or count
#'    type is permitted.
#' @param VarsC chr. Covariates included in mediation modelling.
#'    When "default" is specified, all covariates in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "C1,C2".
#' @param VarsM chr. Mediator variables included in pairwise mediation modelling.
#'    When "default" is specified, all mediator variables in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "M1,M2,M3".
#' @param Method chr. Dimension reduction method. Available options include
#'    "gcdnet" and "mean"(default). "mean" option is recommended when outcome variable
#'    is a binary variable because in some situations low variance of shrinkaged exposure
#'    variable might obtain in "gcdnet" settings for binary outcome.
#' @param Folds num. Number of cross validation for gcdnet method. Default is 10.
#' @param Family chr. The error distribution and link function to be used in the
#'    model. Available options include "linear" and "gaussian" for continuous
#'    outcome variables, and "binomial" as well as "poisson" options for binary
#'    and count outcome variables.
#' @param Iter num. The number of iteration times. Default is 500. To get a more stable
#'    result, a minimum iteration of 1,000 is recommended. To obtain more digits
#'    of P value of parameter estimation, a minimum iteration of 5,000 is suggested.
#' @details RedX function provides two alternative methods for exposure dimension
#'    reduction, including sum method as well as adaptive elastic net method via
#'    gcdnet package. By default, all exposures will be included for dimension reduction.
#'    Afterwards, mediation models will be built between given mediators as well as
#'    shrinkaged exposure variables.
#' @return A list containing three elements where the exposure dimension reduction
#'    variables as well as their mediation modelling results with mediators
#'    are stored. That list include:
#'    1. "MedtRedX_ERSall": a dataframe containing the exposure dimension reduction
#'    result where all exposures are considered as one group (ERS_All).
#'    2. "MedtRedX_ERSlist": a dataframe containing the exposure dimension reduction
#'    results where exposures are shrinkaged in their own subgroups.
#'    3. "MedtRedX_Stats": a dataframe containing the mediation modelling results
#'    where the shrinkaged exposure variables were paired with each mediator provided.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res4 <- RedX(PID=res$PID, VarsY = "Y1",
#'    VarsC = "C1", VarsM = "M1,M2,M3", Method = "mean", Folds = 10,
#'    Family = "linear",Iter = 500)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

RedX = function(PID,
                OutPath = "default",
                VarsY = "Y1",
                VarsC = "default",
                VarsM = "default",
                Method = "mean",
                Folds = 10,
                Family = "linear",
                Iter = 500){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'RedX')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MedtRedXVarsY=VarsY,
                               MedtRedXVarsC=VarsC,
                               MedtRedXVarsM=VarsM,
                               MedtRedXMethod=Method,
                               MedtRedXFolds=Folds,
                               MedtRedXFamily=Family,
                               MedtRedXIter=Iter
                   ),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/2.1-RedX")
  if(!file.exists(path)){dir.create(path)}
  vroom::vroom_write(results$MedtRedX_Stats,paste0(path, "/RedX.modelling.results.csv"),",")
  try(vroom::vroom_write(results$MedtRedX_ERSall,paste0(path, "/RedX.MedtRedX_ERSall.csv"),","),silent=TRUE)
  try(vroom::vroom_write(results$MedtRedX_ERSlist,paste0(path, "/RedX.MedtRedX_ERSlist.csv"),","),silent=TRUE)
  return(results)
}


#' @title Mediator dimension reduction
#' @description Implement mediator dimension reduction for mediators.
#' @usage RedM(PID, OutPath, VarsY, Family, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default",
#'    the current working directory will be set.It should be noted that the
#'    slash symbol is "/", not "\".
#' @param VarsY chr. The outcome variable. Either continuous, binary or count
#'    type is permitted.
#' @param VarsX chr. Exposure variables included in mediation modelling.
#'    When "default"(recommended) is specified, all exposure variables in the data
#'    will be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "X1,X2,X3".
#' @param VarsC chr. Covariates included in mediation modelling.
#'    When "default" is specified, all covariates in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "C1,C2".
#' @param Method chr. Dimension reduction method. Available options include
#'    "mean" and "pdm1"(default). "mean" option is recommended when outcome variable
#'    is a binary variable because in some situations error might occur in "pdm1"
#'    settings for binary outcome.
#' @param Family chr. The error distribution and link function to be used in the
#'    model. Available options include "linear" and "gaussian" for continuous
#'    outcome variables, and "binomial" as well as "poisson" options for binary
#'    and count outcome variables.
#' @param Iter num. The number of iteration times. Default is 500. To get a more stable
#'    result, a minimum iteration of 1,000 is recommended. To obtain more digits
#'    of P value of parameter estimation, a minimum iteration of 5,000 is suggested.
#' @details RedM function provides two alternative methods for mediator dimension
#'    reduction, including sum method as well as pdm1 method via
#'    PDM package. By default, all mediators will be included for dimension reduction.
#'    Afterwards, mediation models will be built between given exposures as well as
#'    shrinkaged mediator variables.
#' @return A list containing two elements where the mediator dimension reduction
#'    modelling results with exposures are stored. That list include:
#'    1. "MedtRedM_all": a dataframe containing the mediator dimension reduction
#'    result where all mediators are considered as one group and that shrinkaged
#'    mediator is paired and modelled with each exposure.
#'    2. "MedtRedM_list": a dataframe containing the mediator dimension reduction
#'    result where mediators are shrinkaged in their own subgroups, and these
#'    shrinkaged mediators are paired and modelled with each exposure.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res4 <- RedM(PID=res$PID, VarsY = "Y1", VarsX = "X1,X2,X3",
#'    VarsC = "C1", Method = "mean", Family = "linear", Iter = 500)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

RedM = function(PID,
                OutPath = "default",
                VarsY = "Y1",
                VarsX = "default",
                VarsC = "default",
                Method = "mean",
                Family = "linear",
                Iter = 500){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'RedM')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MedtRedMVarsY=VarsY,
                               MedtRedMVarsX=VarsX,
                               MedtRedMVarsC=VarsC,
                               MedtRedMMethod=Method,
                               MedtRedMFamily=Family,
                               MedtRedMIter=Iter
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/2.2-RedM")
  if(!file.exists(path)){dir.create(path)}
  try(vroom::vroom_write(results$MedtRedM_all,paste0(path, "/RedM-all.mediators.csv"),","),silent=TRUE)
  try(vroom::vroom_write(results$MedtRedM_list,paste0(path, "/RedM-each.mediator.group.csv"),","),silent=TRUE)
  return(results)
}


#' @title Exposure and mediator dimension reduction
#' @description Implement exposure and mediator dimension reduction. It should be
#'    noted that the exposure dimension reduction result has been built by
#'    RedX function prior to using it.
#' @usage RedXM(PID, OutPath, VarsY, Family, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default",
#'    the current working directory will be set.It should be noted that the
#'    slash symbol is "/", not "\".
#' @param VarsY chr. The outcome variable. Either continuous, binary or count
#'    type is permitted.
#' @param VarsC chr. Covariates included in mediation modelling.
#'    When "default" is specified, all covariates in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "C1,C2".
#' @param Method chr. Dimension reduction method. Available options include
#'    "mean" and "pdm1"(default). "mean" option is recommended when outcome variable
#'    is a binary variable because in some situations error might occur in "pdm1"
#'    settings for binary outcome.
#' @param Family chr. The error distribution and link function to be used in the
#'    model. Available options include "linear" and "gaussian" for continuous
#'    outcome variables, and "binomial" as well as "poisson" options for binary
#'    and count outcome variables.
#' @param Iter num. The number of iteration times. Default is 500. To get a more stable
#'    result, a minimum iteration of 1,000 is recommended. To obtain more digits
#'    of P value of parameter estimation, a minimum iteration of 5,000 is suggested.
#' @details RedXM function provides two alternative methods for mediator dimension
#'    reduction, including sum method as well as pdm1 method via PDM package.
#'    By default, all mediators will be included for dimension reduction.
#'    Afterwards, mediation models will be built between shrinkaged exposure variables
#'    as well as shrinkaged mediator variables. Prior to using RedXM, make sure that
#'    RedX function has been executed to obtain shrinkaged exposure variables.
#' @return A list containing two elements where the mediator dimension reduction
#'    modelling results with shrinkaged exposures are stored. That list include:
#'    1. "MedtRedXM_all": a dataframe containing the mediator dimension reduction
#'    result where all mediators are considered as one group and that shrinkaged
#'    mediator is paired and modelled with shrinkaged exposures.
#'    2. "MedtRedXM_list": a dataframe containing the mediator dimension reduction
#'    result where mediators are shrinkaged in their own subgroups, and these
#'    shrinkaged mediators are paired and modelled with shrinkaged exposures.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res3 <- RedX(PID=res$PID, VarsY = "Y1",
#'    VarsC = "default", VarsM = "default", Method = "mean", Folds = 10,
#'    Family = "linear", Iter = 500)
#'    res4 <- RedXM(PID=res$PID, VarsY = "Y1",
#'    VarsC = "default", Method = "mean", Family = "linear", Iter = 500)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

RedXM = function(PID,
                 OutPath = "default",
                 VarsY = "Y1",
                 VarsC = "default",
                 Method = "mean",
                 Family = "linear",
                 Iter = 500){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'RedXM')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MedtRedXMVarsY=VarsY,
                               MedtRedXMVarsC=VarsC,
                               MedtRedXMMethod=Method,
                               MedtRedXMFamily=Family,
                               MedtRedXMIter=Iter
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/2.3-RedXM")
  if(!file.exists(path)){dir.create(path)}
  try(vroom::vroom_write(results$MedtRedXM_all,paste0(path, "/RedXM-all.mediators.csv"),","),silent=TRUE)
  try(vroom::vroom_write(results$MedtRedXM_list,paste0(path, "/RedXM-each.mediator.group.csv"),","),silent=TRUE)
  return(results)
}


#' @title Visualize RedXM mediation modelling result
#' @description Visualize the RedXM mediation result. It should be noted that
#'    the RedXM mediation modelling result has been built by RedXM function
#'    prior to using it.
#' @usage VizRedXM(PID, OutPath, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default",
#'    the current working directory will be set.It should be noted that the
#'    slash symbol is "/", not "\".
#' @param Brightness chr. Visualization brightness. Available options include
#'    "Light" and "Dark".
#' @param Pallette chr. Visualization palette. Available options include "default1",
#'    "default2" and "Journal". The "Journal" option provides several journal
#'    preference styles including cell, nature, science, lancet, nejm, and jama.
#' @details VizRedXM draws a visualized display for RedXM results.
#'    Piror to using VizMedtPair, make sure that the users have built mediation
#'    models by RedXM function.
#' @return A list containing two elements where the visualized exposure and
#'    mediator dimension reduction plot as well as the organized plotting data
#'    are stored. The elements of that list include:
#'    1. "MedtRedXM_plotadta": a dataframe containing the organized plotting
#'    data extracted from exposure and mediator dimension reduction result.
#'    2. "MedtRedXM_plot": a visualized plot for exposure and mediator
#'    dimension reduction result.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res3 <- RedX(PID=res$PID, VarsY = "Y1",
#'    VarsC = "default", VarsM = "default", Method = "mean", Folds = 10,
#'    Family = "linear", Iter = 500)
#'    res4 <- RedXM(PID=res$PID, VarsY = "Y1",
#'    VarsC = "default", Method = "mean", Family = "linear", Iter = 500)
#'    res5 <- VizRedXM(PID=res$PID, Brightness = "light",
#'    Pallette = "nature")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

VizRedXM = function(PID,
                    OutPath = "default",
                    Brightness = "light",
                    Pallette = "default1"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizRedXM')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizRedXMBrightness=Brightness,
                               VizRedXMPallette=Pallette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/2.3-RedXM")
  if(!file.exists(path)){dir.create(path)}
  vroom::vroom_write(results$MedtRedXM_plotadta,paste0(path, "/VizMedtRedXM.plotdata.csv"),",")

  ggplot2::ggsave(plot = results$MedtRedXM_plot,
                  filename = stringr::str_c("/RedXM-forest.plot-",Brightness,".",Pallette,"style",".jpg"),
                  path = path,
                  width = 36,
                  height = 18,
                  units = "cm")
  return(results)
}


#' @title Estimate the importance of mediators
#' @description Estimate the importance of mediators.
#' @usage ImptM(PID, OutPath, VarsY, ...)
#' @param PID chr. Program ID. It must be the same with the PID generated
#'    by InitStatMedt.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default",
#'    the current working directory will be set.It should be noted that the
#'    slash symbol is "/", not "\".
#' @param VarsY chr. The outcome variable. Either continuous, binary or count
#'    type is permitted.
#' @param VarsX chr. Exposure variables included in estimation procedure.
#'    When "default" is specified, all exposure variables in the data
#'    will be used. The shirnkaged exposure variables are also permitted.
#'    It should be noted that there is fixed format for the characters separated
#'    with comma and without space, e.g., "X1,X2,X3".
#' @param VarsC chr. Covariates included in estimation procedure.
#'    When "default" is specified, all covariates in the data will
#'    be used. It should be noted that there is fixed format for the characters
#'    separated with comma and without space, e.g., "C1,C2".
#' @details ImptM function estimates the importance of mediators among given exposures.
#'    The estimation procedure is mainly realized by the hima function in HIMA package.
#'    ImptM also provides another evaluation method by the bama function in bama package.
#'    The users can search for these packages for further information.
#' @return A list containing four elements where the raw estimation results
#'    as well as the tidy tables for display are stored. The elements of
#'    that list include:
#'    1. "MedtImptM_all": the raw estimation result where the importance of
#'    each mediator was evaluated among whole mediators.
#'    2. "MedtImptM_list": the raw estimation result where the importance of
#'    each mediator was evaluated among each mediator group.
#'    3. "MedtImptM_table1": the tidy table for MedtImptM_all and MedtImptM_list,
#'    where significant estimations (q value <0.2) are expressed with an asterisk
#'    mark (*).
#'    4. "MedtImptM_table2": the tidy table for MedtImptM_all and MedtImptM_list,
#'    where only significant estimations (q value <0.2) are displayed in this table.
#'    MedtImptM_table2 is as same content as MedtImptM_table1.
#' @export
#' @examples res <- InitStatMedt()
#'    res1 <- LoadStatMedt(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 = TidyStatMedt(PID = res$PID, OutPath = "default",LogTransM = "log10",
#'    RangeLow = 0,RangeUpp = 1)
#'    res3 <- RedX(PID=res$PID, VarsY = "Y1",
#'    VarsC = "default", VarsM = "default", Method = "mean", Folds = 10,
#'    Family = "linear", Iter = 500)
#'    res4 <- ImptM(PID = res$PID, VarsY = "Y1",
#'    VarsX = "default", VarsC = "default")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

ImptM = function(PID,
                 OutPath = "default",
                 VarsY = "Y1",
                 VarsX = "default",
                 VarsC = "default"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'ImptM')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MedtImptMVarsY=VarsY,
                               MedtImptMVarsX=VarsX,
                               MedtImptMVarsC=VarsC
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/2.4-ImptM")
  if(!file.exists(path)){dir.create(path)}
  try(vroom::vroom_write(results$MedtImptM_all,paste0(path, "/ImptM-all.mediators.csv"),","),silent=TRUE)
  try(vroom::vroom_write(results$MedtImptM_list,paste0(path, "/ImptM-each.mediator.group.csv"),","),silent=TRUE)
  try(vroom::vroom_write(results$MedtImptM_table1,paste0(path, "/ImptM-table1.csv"),","),silent=TRUE)
  try(vroom::vroom_write(results$MedtImptM_table2,paste0(path, "/ImptM-table2.csv"),","),silent=TRUE)
  return(results)
}

#' @title Initialize StatSurv module
#' @description Initialize StatSurv module analysis. It can generate an R6
#'     class object integrating all the analysis information.
#' @usage InitStatSurv()
#' @details InitStatSurv uses R6 package to generate an R6 class object where
#'    parameters to be used for the following mediation module program are
#'    initialized and save in that object. All executed function codes in the
#'    StatSurv packaged will be recorded in the form of log text in that
#'    object. Furthermore, a program ID (i.e., PID) is randomly created for the
#'    users to identify their own program.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatSurv()
#'    FuncExit(PID = res$PID)
#' @author Changxin Lan, Bin Wang(corresponding author)
InitStatSurv = function(){
  url = paste0(urlhead,'InitSurv')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}

#' @title Load data file for Survival module
#' @description Load data file for Survival module
#' @usage LoadStatSurv(PID = PID, UseExample = "default", ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by
#'    InitStatSurv.
#' @param UseExample chr. Method of uploading data. If "default",
#'    user should upload their own data files, or use "example#1"
#'    provided by this module.
#' @param DataPath chr. Input data file directory, e.g.
#'    "D:/test/eg_Surv_data.xlsx". It should be noted that the slash
#'    symbol is "/", not "\".
#' @param VocaPath chr. Input vocabulary file directory, e.g.
#'    "D:/test/eg_Surv_voca.xlsx". It should be noted that the slash
#'    symbol is "/", not "\".
#' @details LoadStatSurv function loads the data file and the vocabulary file into
#'    the R6 object that InitStatSurv created. Noted that there are several data
#'    format requirments for the data and vocabulary file.
#'    For data file,the first three columns should be named as "SampleID", "SubjectID", and "Group", respectively.
#'    For the "Group" variable, only two values can be used, i.e. "train" and "test". If there is no data for test, all values should be set as "train".
#'    For outcome variables, their initials must be set as "Y" and serialized by adding Arabic numerals if needed, e.g., Y1, Y2, Y3. In this module, the survival time (Y1) and status (Y2) must be provided.For exposure variables, their initials must be set as "X" and serialized by adding Arabic numerals if needed, e.g., X1, X2, X3.
#'    For covariate variables, their initials must be set as "C" and serialized by adding Arabic numerals if needed, e.g., C1, C2, C3. It should be noted the covariates are not required if users don’t have.
#'    For vocabulary file, the first two columns must be named as "SerialNo" and "FullName", respectively.
#'    The list of SerialNo of outcomes, exposure, and covariates should be the same with the column names of "Data file".
#'    The list of the FullName is prepared as users’ like.

#' @return An R6 class object containing the input data and vocabulary file.
#' @export
#' @examples res <- InitStatSurv()
#'    res1 <- LoadStatSurv(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    FuncExit(PID = res$PID)
#' @author Changxin Lan, Bin Wang(corresponding author)
LoadStatSurv =function(PID,
                       UseExample="default",
                       DataPath=NULL,
                       VocaPath=NULL){
  url = paste0(urlhead,'LoadSurv')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata,
                                         Vocadata=vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Transform factor variables into dummy ones
#' @description Transform factor variables into dummy ones
#' @usage TidyStatSurv(PID, OutPath = "default",TransDummyVars="default")
#' @param PID chr. Program ID. It must be the same with the PID generated by ExpoSurv
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TransDummyVars chr. Variables to be transformed as dummy variables. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., "X1,X2,X3". If "default", all the factor variables will be transformed into dummy ones.
#'     These variables need to be transformed as factor ones in previous transform step
#'     using TransType function.
#' @return An R6 class object containing the variable(s) after transforming the factor
#'     variables into dummy ones.
#' @export
#' @examples res <- InitStatSurv()
#'     res1 = LoadStatSurv(PID = res$PID, UseExample = "example#1")
#'     res2 = TidyStatSurv(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Bin Wang
TidyStatSurv = function(PID,
                        OutPath = "default",
                        TransDummyVars="default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidySurv')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransDummyVars=TransDummyVars),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyStatSurv'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyStatSurv/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyStatSurv/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Association analysis
#' @description Association analysis for survival data
#' @usage SurvAsso = function(PID, OutPath = "default",
#'     TimeY, EventY,VarsX='all.x',VarsN = "single.factor",VarsSel = F,
#'     Covariates = "all.c")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatSurv()
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TimeY chr. Outcome variable of survival time used for modelling. Only one variable can be entered.
#' @param EventY chr. Outcome variable of status used for modelling. Only one variable can be entered.
#' @param VarsX Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param VarsSel lgl. T (or TRUE) and F (or FALSE). Whether to select the significant variable for the final model.
#'     Available options includes T and F
#' @param Covariates chr. Covariates used for modeling. The default option is
#'     "all.c" (All covariates are included).
#'
#' @return A list containing the association analysis results.
#' @export
#'
#' @examples res <- InitStatSurv()
#'     res1 = LoadStatSurv(PID = res$PID, UseExample = "example#1")
#'     res2 = TidyStatSurv(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     res3 = SurvAsso(PID=res$PID, TimeY = "Y1", EventY= 'Y2',VarsX='all.x',
#'     VarsN="single.factor",VarsSel="T",Covariates = "all.c")
#'     FuncExit(PID = res$PID)
#' @author Changxin Lan, Bin Wang(corresponding author)
SurvAsso = function(PID,
                    OutPath='default',
                    TimeY,
                    EventY,
                    VarsX='all.x',
                    VarsN="single.factor",
                    VarsSel=T,
                    Covariates = "all.c"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'SurvAsso')
  res = httr::POST(url,
                   body = list(PID=PID,
                               SurvAssoTimeY=TimeY,
                               SurvAssoEventY=EventY,
                               SurvAssoVarsX=VarsX,
                               SurvAssoVarsN=VarsN,
                               SurvAssoVarsSel=VarsSel,
                               SurvAssoCova=Covariates),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/Association")
  if(!file.exists(path)){dir.create(path)}
  vroom::vroom_write(results$coxph_res,paste0(path, "/", EventY,"_",VarsN,".csv"),",")
  return(results)
}

#' @title Visualize association analysis
#' @description Visualize association analysis
#' @usage VizSurvAsso = function(PID, OutPath = "default",VarsN = "single.factor",
#'     Layout = "volcano",Brightness = "light",Palette = "default1",ColorFor= "p.value",SizeFor= "p.value")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatSurv()
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param Layout chr. Visualization layout. Available options include "forest" and "volcano".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., "cell", "nature", "science", "lancet", "nejm", and "jama").
#' @param ColorFor chr. Volcano plot dot color. Available options include "p.value" and "hr".
#' @param SizeFor chr. Volcano plot dot size. Available options include "p.value" and "hr".
#'
#' @return A list containing the results' plot.
#' @export
#'
#' @examples res <- InitStatSurv()
#'    res1 = LoadStatSurv(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatSurv(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = SurvAsso(PID=res$PID, TimeY = "Y1", EventY= 'Y2',VarsX='all.x',
#'    VarsN="single.factor",VarsSel="T",Covariates = "all.c")
#'    res4 = VizSurvAsso(PID=res$PID,VarsN="single.factor",Layout="volcano",Brightness= "light",
#'    Palette = "default1",ColorFor= "p.value",SizeFor= "p.value")
#'    FuncExit(PID = res$PID)
#' @author Changxin Lan, Ning Gao, Bin Wang(corresponding author)


VizSurvAsso = function(PID,
                       OutPath='default',
                       VarsN="single.factor",
                       Layout="volcano",
                       Brightness= "light",
                       Palette = "default1",
                       ColorFor= "p.value",
                       SizeFor= "p.value"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra,GGally)
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'VizSurvAsso')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizSurvAssoVarsN=VarsN,
                               VizSurvAssoLayout =Layout,
                               VizSurvAssoBrightness = Brightness,
                               VizSurvAssoPalette = Palette,
                               VizSurvAssoColorFor = ColorFor,
                               VizSurvAssoSizeFor =SizeFor),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/Association")
  if (VarsN=="single.factor"){
    for (i in c(1:length(results$SurvCurvePlot))){
      ggplot2::ggsave(paste0(path, "/", names(results$SurvCurvePlot)[i], "_SingleFactor_SurvCurve.png"),
                      plot = results$SurvCurvePlot[[i]],
                      width = 35,
                      height = 20,
                      units = "cm")
    }
  }else{
    ggplot2::ggsave(paste0(path, "/MultipleFactor_SurvCurve.png"),
                    plot = results$SurvCurvePlot,
                    width = 35,
                    height = 20,
                    units = "cm")
  }
  if(Layout =="forest"){
    ggplot2::ggsave(paste0(path, "/",VarsN,"_forest_",Brightness,"_",Palette,".png"),
                    plot = results$ForestPlot,
                    width = 15,
                    height = 5+10*ceiling(results$nrow_forest/50),
                    units = "cm",dpi = 300,
                    limitsize = FALSE)
  }
  if (Layout =="volcano"){
    ggplot2::ggsave(paste0(path, "/",VarsN,"_volcano_",Brightness,"_",Palette,"_colorforhr_sizeforpvalue.png"),
                    plot = results$VolcanoPlot,
                    width = 20+5*ceiling(results$nrow_vocanol/50),
                    height = 9+5*ceiling(results$nrow_vocanol/50),
                    units = "cm",dpi = 300)
  }
  return(results)
}


#' @title Build prediction models
#' @description Build prediction models
#' @usage SurvPred = function(PID,OutPath = "default",TimeY,EventY,VarsX = "all.x",
#'     Covariates = "all.c",RsmpMethod = "cv",Folds = 3,Ratio = 0.667,Repeats = 3)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatSurv()
#' @param OutPath  chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TimeY chr. Outcome variable of survival time used for modelling. Only one variable can be entered.
#' @param EventY chr. Outcome variable of status used for modelling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param Covariates chr. Covariates used for modeling. The default option is
#'     "all.c" (All covariates are included).
#' @param RsmpMethod chr. Three resampling methods options for internal validation, including
#'     "cv" (i.e.,Cross validation) , "bootstrap", and "holdout".
#' @param Folds num. Folds of Cross-validation resampling. It is ranging 2-10.
#' @param Ratio num. Ratio of Bootstrap resampling. It is ranging 0.4-0.9.
#' @param Repeats num. Number of Bootstrap resampling. It is ranging 2-20.
#'
#' @return  A list containing the prediction performance evaluation.
#' @export
#'
#' @examples res <- InitStatSurv()
#'    res1 = LoadStatSurv(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatSurv(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = SurvPred(PID=res$PID, TimeY = "Y1", EventY= 'Y2',VarsX='all.x',
#'    Covariates = "all.c",RsmpMethod="cv",Folds=3,Ratio=0.667,Repeats=3)
#'    FuncExit(PID = res$PID)
#' @author Changxin Lan, Bin Wang(corresponding author)
SurvPred = function(PID,
                    OutPath='default',
                    TimeY,
                    EventY,
                    VarsX= "all.x",
                    Covariates = "all.c",
                    RsmpMethod="cv",
                    Folds=3,
                    Ratio=0.667,
                    Repeats=3){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'SurvPred')
  res = httr::POST(url,
                   body = list(PID=PID,
                               SurvPredTimeY=TimeY,
                               SurvPredEventY=EventY,
                               SurvPredVarsX=VarsX,
                               SurvPredCova=Covariates,
                               SurvPredRsmpMethod=RsmpMethod,
                               SurvPredFolds=Folds,
                               SurvPredRatio=Ratio,
                               SurvPredRepeats=Repeats),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/Prediction")
  if(!file.exists(path)){dir.create(path)}
  vroom::vroom_write(results$bmrout,paste0(path, "/",EventY, "_BMR_", RsmpMethod,".csv"),",")
  for (i in c(1:length(results$predout))){
    try(vroom::vroom_write(data.frame(results$predout[[i]]),paste0(path, "/",EventY, "_Prediton_",
                                                                   names(results$predout)[i],".csv"),","),silent = TRUE)
  }
  return(results)
}

#' @title Visualize the prediction performance
#' @description Visualize the prediction performance
#' @usage VizSurvPred = function(PID,OutPath = "default",Layout = "curve",
#'     Brightness = "light",Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatSurv()
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param Layout chr. Visualization layout. Available options include "curve" , "bar" and 'all'.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., "cell", "nature", "science", "lancet", "nejm", and "jama").
#'
#' @return A list containing the results' plot.
#' @export
#'
#' @examples res <- InitStatSurv()
#'    res1 = LoadStatSurv(PID = res$PID, UseExample = "example#1")
#'    res2 = TidyStatSurv(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = SurvPred(PID=res$PID, TimeY = "Y1", EventY= 'Y2',VarsX='all.x',
#'    Covariates = "all.c",RsmpMethod="cv",Folds=3,Ratio=0.667,Repeats=3)
#'    res4 = VizSurvPred(PID=res$PID,Layout="curve",Brightness="light",Palette='default1')
#'    FuncExit(PID = res$PID)
#' @author Changxin Lan, Ning Gao, Bin Wang(corresponding author)
VizSurvPred = function(PID,
                       OutPath='default',
                       Layout="curve",
                       Brightness="light",
                       Palette='default1'){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra,GGally)
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'VizSurvPred')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizeSurvPredLayout =Layout,
                               VizeSurvPredBrightness =Brightness,
                               VizeSurvPredPalette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/Prediction")
  if(!file.exists(path)){dir.create(path)}
  try(
    ggplot2::ggsave(paste0(path, "/learner_fit_survcurve_comp_",Brightness,"_",Palette,".png"),
                    plot = results$PredSurvCurvPlot,
                    width = 30,
                    height = 20,
                    units = "cm"),
    silent = TRUE
  )
  try(
    ggplot2::ggsave(paste0(path, "/BMR_barplot_",Brightness,"_",Palette,".png"),
                    plot =results$BmrBarPlot,
                    width = 25,
                    height = 15,
                    units = "cm",dpi = 300),
    silent = TRUE
  )
  try(
    ggplot2::ggsave(paste0(path, "/BMR_Pointplot_",Brightness,"_",Palette,".png"),
                    plot = results$BmrPointPlot,
                    width = 25,
                    height = 15,
                    units = "cm",dpi = 300)
  )
  return(results)
}

#' @title Compare the survival curves of two groups
#' @description Compare the survival curves of two groups
#' @usage VizSurvCompGroup = function(PID,OutPath = "default",TimeY,EventY,
#'     VarsG,Model='km',VarsAdj,AdjMethod='average',Brightness = "light",Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatSurv()
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TimeY chr. Outcome variable of survival time used for modelling. Only one variable can be entered.
#' @param EventY chr. Outcome variable of status used for modelling. Only one variable can be entered.
#' @param VarsG chr.  Grouping variable, must be a binary variable.
#' @param Model chr. Methods to depict the survival curve. Options include 'km' (Kaplan-Meier estimate) and
#'     "coxph" (Cox proportional hazards regression mode).
#' @param VarsAdj If you choose the cox model, co-variables used for modelling.
#'     It should be noted that there is fixed format for the entering characters separated with comma and without space, e.g., X1,X2,X3.
#' @param AdjMethod If you choose the cox model, method for adjusting model, include: "average","single","margin" and "conditional".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., "cell", "nature", "science", "lancet", "nejm", and "jama").
#'
#' @return A list containing the results' plot.
#' @export
#'
#' @examples res <- InitStatSurv()
#'    res1 = LoadStatSurv(PID = res$PID, UseExample = "example#1")
#'    res2 = VizSurvCompGroup(PID=res$PID,TimeY="Y1",EventY="Y2",VarsG="C3",
#'    Model="km",Brightness="light",Palette='default1')
#'    FuncExit(PID = res$PID)
#' @author Changxin Lan, Ning Gao, Bin Wang(corresponding author)
VizSurvCompGroup = function(PID,
                            OutPath='default',
                            TimeY,
                            EventY,
                            VarsG,
                            Model='km',
                            VarsAdj='',
                            AdjMethod='average',
                            Brightness='light',
                            Palette='default1'){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  message("It may take some time. Thanks for your patient waiting!")
  url = paste0(urlhead,'VizSurvCompGroup')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizSurvCompGroupTimeY=TimeY,
                               VizSurvCompGroupEventY=EventY,
                               VizSurvCompGroupVarsG=VarsG,
                               VizSurvCompGroupVarsAdj=VarsAdj,
                               VizSurvCompGroupModel=Model,
                               VizSurvCompGroupAdjMethod=AdjMethod,
                               VizSurvCompGroupBrightness =Brightness,
                               VizSurvCompGroupPalette =Palette),
                   encode = 'json')

  results = unserialize(httr::content(res))
  path = paste0(OutPath, "/VizSurvCompGroup")
  if(!file.exists(path)){dir.create(path)}
  if (!is.null(results$Comp_KM_Plot)){
    for (i in c(1:length(results$Comp_KM_Plot))){
      ggplot2::ggsave(paste0(path, "/km", "_",EventY,"_",
                             names(results$Comp_KM_Plot)[i],"_",
                             Brightness,"_",Palette,".png"),
                      plot = results$Comp_KM_Plot[[i]],
                      width = 35,
                      height = 60,
                      units = "cm"
      )
    }
  }

  if (!is.null(results$Comp_Coxph_Plot)){
    for (i in c(1:length(results$Comp_Coxph_Plot))){
      ggplot2::ggsave(paste0(path, "/coxph_", EventY, "_",
                             names(results$Comp_Coxph_Plot)[i],"_",
                             AdjMethod,"_", Brightness,"_",Palette,".png"),
                      plot = results$Comp_Coxph_Plot[[i]],
                      width = 35,
                      height = 20,
                      units = "cm",dpi = 300)
    }
  }
  return(results)
}


#' @title Initialize EMS module
#' @description Initialize EMS module analysis. It can generate an R6 class object.
#' @usage InitEMS()
#' @param
#' @details ExpoNontarget module is designed to conduct the analysis of the features
#'     from the high-resolution mass spectrometry. It mainly aims to screen and
#'     annotate the significant features associated with the health outcomes.
#' @return An R6 class object.
#' @export
#' @examples res <- InitEMS()
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)

InitEMS = function(){
  url = paste0(urlhead,'InitNTA')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for EMS module
#' @description Load data file for EMS module
#' @usage LoadEMS(PID, UseExample = "default", MS2, DataPath=NULL, VocaPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by EMS
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1","example#2" provided by this module.
#' @param MS2 lgl.Whether to use MS2 information to annotate the target ion. If use example data,
#'    choose "T" to use "example#2" data with MS2 information, choose "F" to use "example#1" data
#'    without MS2 information.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input directory of vocabulary file, e.g. "D:/test/eg_voca_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitEMS()
#'    res1 = LoadEMS(PID = res$PID, UseExample = "example#1", MS2 = FALSE)
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)

LoadEMS =function(PID,
                  UseExample="default",
                  MS2 = F,
                  DataPath=NULL,
                  VocaPath=NULL,
                  Mzmldata=NULL){
  url = paste0(urlhead,'LoadNTA')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         MS2=MS2,
                                         Expodata=exdata,
                                         Vocadata=vodata,
                                         Mzmldata=Mzmldata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Tidy data.
#' @description Tidy data file for EMS module
#' @usage TidyEMS(PID, OutPath = "default",TransDummyVars="default")
#' @param PID chr. Program ID. It must be the same with the PID generated by EMS
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param TransDummyVars chr. Variables to be transformed as dummy variables. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., "X1,X2,X3". If "default", all the factor variables will be transformed into dummy ones.
#' @return An R6 class object containing the variable(s) after transforming the factor
#'     variables into dummy ones.
#' @export
#' @examples res <- InitEMS()
#'     res1 = LoadEMS(PID = res$PID, UseExample = "example#1", MS2 = FALSE)
#'     res2 = TidyEMS(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'     FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)
TidyEMS = function(PID,
                   OutPath = "default",
                   TransDummyVars="default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyNTA')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransDummyVars=TransDummyVars),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyEMS'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyEMS/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyEMS/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}


#' @title Association analysis
#' @description Association analysis for non-targeted data
#' @usage NtaCros(PID=res$PID, OutPath="default",VarsY = "Y1", VarsX = "all.x",
#'    VarsN = "single.factor", FdrCorrect = "T", SelMethod = "lasso",
#'    StepwizeThr = 0.1, RF_ImpThr = 0.9, Covariates = "all.c", Family = "gaussian",
#'    RepMsr = "F", Corstr = "ar1")
#' @param PID chr. Program ID. It must be the same with the PID generated by EMS
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param FdrCorrect lgl. T (or TRUE) and F (or FALSE).  Whether to correct the multiple hypotheses by false positive rate (FDR) method.
#' @param SelMethod chr. Methods to select the significant features. Options include
#'     "stepwise" (multiple linear regress using stepwise algorithm),
#'     "lasso" (multiple linear regress using LASSO regularization algorithm),
#'     "random.forest" (random forest).
#' @param StepwizeThr num. Threshold of the P value for stepwise regression to
#'     screen important variables. It ranges 0.05-0.25 with the default value of 0.1.
#' @param RF_ImpThr num. Threshold of the total importance for the variables to a random forest model. It ranges 0.5-1.0 with the default value of 0.9.
#' @param Covariates chr. Covariates used for modeling. It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g. "X1,X2,X3".The default option is"all.c" (All covariates are included).
#' @param Family chr. The link function for the regression model according
#'     the data type of outcomes, including "gaussian" for continuous variable,
#'     "binomial" for binary variable, and "poisson" for counting variable
#' @param RepMsr lgl. T (or TRUE) and F (or FALSE). Whether existing repeated measurement of the subjects.
#' @param Corstr chr. If "RepMsr" = T, the generalized estimating equations (GEE) will be used.
#'     For GEE, three correlation structure options are "exchangeable" "ar1" "unstructured".
#' @details
#' @return A list containing non-target analysis association results
#' @export
#' @examples res <- InitEMS()
#'    res1 = LoadEMS(PID = res$PID, UseExample = "example#1", MS2 = FALSE)
#'    res2 = TidyEMS(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = NtaCros(PID=res$PID, OutPath="default",VarsY = "Y1", VarsX = "all.x",
#'    VarsN = "single.factor", FdrCorrect = "T", SelMethod = "lasso",
#'    StepwizeThr = 0.1, RF_ImpThr = 0.9, Covariates = "all.c", Family = "gaussian",
#'    RepMsr = "F", Corstr = "ar1")
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)

NtaCros = function(PID,
                   OutPath = "default",
                   VarsY,
                   VarsX = "all.x",
                   VarsN = "single.factor",
                   FdrCorrect = T,
                   SelMethod = "all",
                   StepwizeThr = 0.1,
                   RF_ImpThr = 0.9,
                   Covariates = "all.c",
                   Family,
                   RepMsr  = F,
                   Corstr= "ar1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'NtaCros')
  res = httr::POST(url,
                   body = list(PID=PID,
                               NtaCrosVarsY = VarsY,
                               NtaCrosVarsX = VarsX,
                               NtaCrosVarsN =  VarsN,
                               NtaCrosFdrCorrect = FdrCorrect,
                               NtaCrosSelMethod = SelMethod,
                               NtaCrosStepwizeThr = StepwizeThr,
                               NtaCrosRF_ImpThr = RF_ImpThr,
                               NtaCrosCovariates = Covariates,
                               NtaCrosFamily = Family,
                               NtaCrosRepMsr = RepMsr,
                               NtaCrosCorstr = Corstr),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/NtaCros'))
  for (name in names(results$CrosTable)){
    try(
      vroom::vroom_write(results$CrosTable[[name]],
                         paste0(OutPath,'/NtaCros/', name,".csv"),
                         delim = ","),
      silent = TRUE
    )
  }
  return(results$CrosTable)
}



#' @title Visualize the results of Association analysis
#' @description Visualize the results of Association analysis for non-targeted data
#' @usage VizNtaCros(PID, OutPath = "default", VarsY,VarsN, Layout = "volcano",
#'     EffectThr = 0.5, Brightness = "light", Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by EMS
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor".
#' @param Layout chr. Visualization layout. Available options include "volcano" and "forest".
#' @param EffectThr num. Threshold of the total importance for the variables to a
#'     random forest model. It ranges 0.5-1.0 with the default value of 0.9.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'     and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing plots of non-target analysis association results
#' @export
#' @examples res <- InitEMS()
#'    res1 = LoadEMS(PID = res$PID, UseExample = "example#1", MS2 = FALSE)
#'    res2 = TidyEMS(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = NtaCros(PID=res$PID, OutPath="default",VarsY = "Y1", VarsX = "all.x",
#'    VarsN = "single.factor", FdrCorrect = "T", SelMethod = "lasso",
#'    StepwizeThr = 0.1, RF_ImpThr = 0.9, Covariates = "all.c", Family = "gaussian",
#'    RepMsr = "F", Corstr = "ar1")
#'    res4 = VizNtaCros(PID=res$PID,OutPath="default",VarsY = "Y1",VarsN = "single.factor",
#'    Layout = "volcano",EffectThr = 0.5,Brightness = "light",Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)

VizNtaCros = function(PID,
                      OutPath = "default",
                      VarsY,
                      VarsN,
                      Layout = "volcano",
                      EffectThr = 0.5,
                      Brightness = "light",
                      Palette = "default1"
){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizNtaCros')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizNtaCrosVarsY = VarsY,
                               VizNtaCrosVarsN = VarsN,
                               VizNtaCrosLayout = Layout,
                               VizNtaCrosEffectThr = EffectThr,
                               VizNtaCrosBrightness = Brightness,
                               VizNtaCrosPalette = Palette),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/NtaCros'))
  for (name in names(results$CrosPlot)){

    try(
      ggplot2::ggsave(paste0(OutPath,'/NtaCros/',name,".png"),
                      plot = results$CrosPlot[[name]],
                      width =  results$CrosPlot_para[[name]][['width']],
                      height = results$CrosPlot_para[[name]][['height']],
                      units = "cm",dpi = 300,
                      limitsize = FALSE),
      silent = TRUE
    )
  }
  return(results$CrosPlot)
}

#' @title Annotate the non-targeted features
#' @description Annotate the non-targeted features
#' @usage NtaAnno(PID=res$PID,OutPath="default",VarsY = "Y1",VarsX = "default",VarsN = "single.factor",
#'    FdrCorrect = "F",AdductPos = "M+H",AdductNeg = "M-H",AccuracyMS1 = 5,MS2 = "F",
#'    AccuracyMS2 = 20,ScoreMS2 = 0.4,DiffRT = 5)
#' @param PID chr. Program ID. It must be the same with the PID generated by EMS
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param FdrCorrect lgl. T (or TRUE) and F (or FALSE). Whether to correct the
#'     multiple hypotheses by false positive rate (FDR) method.
#' @param AdductPos chr. Adducts formed in the positive mode using LC-MS.
#'     The default is "all", i.e., all the adducts in positive mode will be chosen, including
#'     "M+3ACN+2H,M+ACN+H,M+NH4,M+2ACN+2H,M+2H,M+3H,M+3Na,M+2Na-H,M+ACN+Na,M+H+Na,
#'     M+2K-H,M+H+NH4,2M+K,M+K,2M+NH4,2M+Na,M+2Na,M+DMSO+H,M+2ACN+H,M+IsoProp+Na+H,
#'     M+2H+Na,M+ACN+2H,M+H,2M+H,M+CH3OH+H,M+H+2Na,M+Na,2M+ACN+H,2M+ACN+Na,M+IsoProp+H,M+H+K"
#' @param AdductNeg chr. Adducts formed in the negative mode using LC-MS.
#'     The default is "all", i.e., all the adducts in positive mode will be chosen, including
#'     "M+FA-H,M+Hac-H,M+Br,3M-H,2M+Hac-H,M+K-2H,2M+FA-H,M-H,M-H2O-H,M+Na-2H,
#'     M-2H,M+TFA-H,M+Cl,M-3H,2M-H”
#' @param AccuracyMS1 num. Accuracy threshold to match the target compounds of MS1 information.
#'     The unit is ppm. It is usually set ranging 1-20.
#' @param MS2 lgl. T (or TRUE) and F (or FALSE). Whether to use MS2 information to annotate the target ion.
#' @param AccuracyMS2 num. Accuracy threshold to match the target compounds of MS2 information.
#'     The unit is ppm. It is usually set ranging 1-20.
#' @param ScoreMS2 num. The lowest matching score ranging 0-1 using MS2 information. The default is 0.4.
#' @param DiffRT num. Difference of the retention time between the measured and the theoretical prediction.
#'     The unit is second (s). It is usually set ranging 60-120.
#' @details
#' @return A list containing non-target analysis results
#' @export
#' @examples res <- InitEMS()
#'    res1 = LoadEMS(PID = res$PID, UseExample = "example#1", MS2 = FALSE)
#'    res2 = TidyEMS(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = NtaCros(PID=res$PID, OutPath="default",VarsY = "Y1", VarsX = "all.x",
#'    VarsN = "single.factor", FdrCorrect = "T", SelMethod = "lasso",StepwizeThr = 0.1,
#'    RF_ImpThr = 0.9, Covariates = "all.c", Family = "gaussian", RepMsr = "F", Corstr = "ar1")
#'    res4 = NtaAnno(PID=res$PID,OutPath="default",VarsY = "Y1",VarsX = "all.x",VarsN = "single.factor",
#'    FdrCorrect = "F",AdductPos = "M+H",AdductNeg = "M-H",AccuracyMS1 = 5,MS2 = "F",
#'    AccuracyMS2 = 20,ScoreMS2 = 0.4,DiffRT = 5)
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)

NtaAnno = function(PID,
                   OutPath = "default",
                   VarsY,
                   VarsX = "default",
                   VarsN = "single.factor",
                   FdrCorrect = F,
                   AdductPos = "all",
                   AdductNeg = "all",
                   AccuracyMS1 = 5,
                   MS2 = F,
                   AccuracyMS2 = 20,
                   ScoreMS2 = 0.4,
                   DiffRT = 5){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'NtaAnno')
  res = httr::POST(url,
                   body = list(PID=PID,
                               NtaAnnoVarsY = VarsY,
                               NtaAnnoVarsX = VarsX,
                               NtaAnnoVarsN = VarsN,
                               NtaAnnoFdrCorrect = FdrCorrect,
                               NtaAnnoAdductPos = AdductPos,
                               NtaAnnoAdductNeg = AdductNeg,
                               NtaAnnoAccuracyMS1 = AccuracyMS1,
                               NtaAnnoMS2 = MS2,
                               NtaAnnoAccuracyMS2 = AccuracyMS2,
                               NtaAnnoScoreMS2 = ScoreMS2,
                               NtaAnnoDiffRT = DiffRT
                   ),
                   encode = 'json')

  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/NtaAnno'))

  for (name in names(results$AnnoTable)){
    try(
      vroom::vroom_write(results$AnnoTable[[name]],
                         paste0(OutPath,'/NtaAnno/', name,".csv"),
                         delim = ","),
      silent = TRUE
    )
  }
  return(results$AnnoTable)
}



#' @title Visualize annotation results
#' @description Visualize annotation results of non-targeted data
#' @usage VizNtaAnno(PID=res$PID,OutPath= "default",VarsY = "Y1",VarsN = "single.factor",
#'    AccuracyMS1 = 5,MS2 = "F",AccuracyMS2 = 20,DiffRT = 100,Brightness = "light",Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by EMS
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsN chr. Choose the single factor or multiple factor model.
#'     Available options include "single.factor" and "multiple.factor"
#' @param AccuracyMS1 num. Accuracy threshold to match the target compounds of MS1 information.
#'     The unit is ppm. It is usually set ranging 1-20.
#' @param MS2 lgl. T (or TRUE) and F (or FALSE). Whether to use MS2 information to annotate the target ion.
#' @param AccuracyMS2 num. Accuracy threshold to match the target compounds of MS2 information.
#'     The unit is ppm. It is usually set ranging 1-20.
#' @param DiffRT num. Difference of the retention time between the measured and the theoretical prediction.
#'     The unit is second (s). It is usually set ranging 60-120.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing plots of non-target annotation analysis results
#' @export
#' @examples res <- InitEMS()
#'    res1 = LoadEMS(PID = res$PID, UseExample = "example#1", MS2 = "F")
#'    res2 = TidyEMS(PID = res$PID, OutPath = "default",TransDummyVars="default")
#'    res3 = NtaCros(PID=res$PID, OutPath="default",VarsY = "Y1", VarsX = "all.x",
#'    VarsN = "single.factor", FdrCorrect = "T", SelMethod = "lasso",
#'    StepwizeThr = 0.1, RF_ImpThr = 0.9, Covariates = "all.c", Family = "gaussian",
#'    RepMsr = "F", Corstr = "ar1")
#'    res4 = NtaAnno(PID=res$PID,OutPath="default",VarsY = "Y1",VarsX = "all.x",VarsN = "single.factor",
#'    FdrCorrect = "F",AdductPos = "M+H",AdductNeg = "M-H",AccuracyMS1 = 5,MS2 = "F",
#'    AccuracyMS2 = 20,ScoreMS2 = 0.4,DiffRT = 5)
#'    res5 = VizNtaAnno(PID=res$PID,OutPath="default",VarsY = "Y1",VarsN = "single.factor",
#'    AccuracyMS1 = 5,MS2 = "F",AccuracyMS2 = 20,DiffRT = 100,Brightness = "light",Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Bin Wang (corresponding author)

VizNtaAnno = function(PID,
                      OutPath = "default",
                      VarsY,
                      VarsN,
                      AccuracyMS1 = 1,
                      MS2 = F,
                      AccuracyMS2 = 20,
                      DiffRT = 100,
                      Brightness = "light",
                      Palette = "default1"
){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'VizNtaAnno')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VizNtaAnnoVarsY = VarsY,
                               VizNtaAnnoVarsN = VarsN,
                               VizNtaAnnoAccuracyMS1 = AccuracyMS1,
                               VizNtaAnnoMS2 = MS2,
                               VizNtaAnnoAccuracyMS2 = AccuracyMS2,
                               VizNtaAnnoDiffRT = DiffRT,
                               VizNtaAnnoBrightness = Brightness,
                               VizNtaAnnoPalette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/NtaAnno'))
  for (name in names(results$AnnoPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/NtaAnno/',name,".png"),
                      plot = results$AnnoPlot[[name]],
                      width =  15,
                      height = 15,
                      units = "cm",dpi = 300),
      silent = TRUE
    )
  }
  return(results$AnnoPlot)
}




#' @title Initialize EDB module
#' @description Initialize EDB module analysis. It can generate an R6 class object.
#' @usage InitEDB()
#' @param
#' @details EDB module is designed as a convenient tool to explore the data,
#'     as well as facilitating to find the biological relationship between exposure
#'     and diseases from the perspective of bioinformatics. This module adopts
#'     the most frequently-used and authoritative databases,
#'     e.g., T3DB, CTD, ToxCast, StringDB, STITCH, KEGG, and GO.
#' @return An R6 class object.
#' @export
#' @examples res <- InitEDB()
#'     FuncExit(PID = res$PID)
#' @author Tianxiang Wu, Bin Wang (corresponding author)

InitEDB = function(){
  url = paste0(urlhead,'InitDb')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for EDB module
#' @description Load data file for EDB module
#' @usage LoadEDB(PID, UseExample = "default", DataPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by EDB
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitEDB()
#'    res1 = LoadEDB(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Tianxiang Wu, Bin Wang (corresponding author)

LoadEDB =function(PID,
                  UseExample="default",
                  DataPath=NULL){
  url = paste0(urlhead,'LoadDb')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
  }else{
    exdata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Explain keywords
#' @description Explain the keyword in ExposomeX platform
#' @usage EDBNode(PID, OutPath = "default", Class, Nodes)
#' @param PID chr. Program ID. It must be the same with the PID generated by EDB
#' @param Class chr. Choose the search range of the concerned keywords. Options include
#'     “Exposure” (In addition to containing chemicals, some exposure factors may be a mixture of chemicals, e.g. PM2.5, tobacco smoking.),
#'     "Metabolite" (the chemicals used for metabolome analysis in the KEGG database),
#'     "Protein", "Enzyme" (referring in particular to the enzymes in the KEGG database),
#'     "Disease", "GO" (gene ontology), and “Gene".
#' @param Nodes chr. Any keywords to search. "default" means using the values in the
#'     input data file. Users can also enter the keywords here. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., "7440-43-9,OMIM:619217,ARF5,GO:0010942".
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @details
#' @return A data frame
#' @export
#' @examples res <- InitEDB()
#'    res1 = LoadEDB(PID = res$PID, UseExample = "example#1")
#'    res2 = EDBNode(PID=res$PID,OutPath = "default",Class = "GO",Nodes = "default")
#'    FuncExit(PID = res$PID)
#' @author Tianxiang Wu, Bin Wang (corresponding author)

EDBNode = function(PID,
                   OutPath = "default",
                   Class,
                   Nodes){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'ExpoNode')
  res = httr::POST(url,
                   body = list(PID=PID,
                               Class=Class,
                               Nodes= Nodes),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Node'))
  try(
    vroom::vroom_write(results$NodeTable,
                       paste0(OutPath,'/Node/Node_',Class,".csv"), delim = ","),
    silent = TRUE
  )
  return(results$NodeTable)
}



#' @title Find the nexuses between keywords
#' @description Find the nexuses between the keywords in ExposomeX platform.
#'     Nexus direction from LinkFrom (class A) to LinkTo (class B)
#' @usage EDBLink(PID, OutPath = "default", ClassA, ClassB, LinkFrom, LinkTo)
#' @param PID chr. Program ID. It must be the same with the PID generated by EDB
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param ClassA chr. Find the Links between the keywords in ClassA and ClassB.
#'     Options include "Exposure","Protein","GO","Pathway","Gene"
#' @param ClassB chr. Options include "Protein","GO","Disease","Pathway","Gene"
#' @param LinkFrom chr. Name, alias, and ID of exposure, protein, GO, gene and pathway are all included.
#'        "Default" means using the values in the input data file.
#' @param LinkTo chr. Name, alias, and ID of protein, GO, gene, pathwayand disease are all included.
#'        "Default" means using the values in the input data file.
#' @details
#' @return A data frame
#' @export
#' @examples res <- InitEDB()
#'    res1 = LoadEDB(PID = res$PID, UseExample = "example#1")
#'    res2 = EDBLink(PID=res$PID,OutPath ="default",ClassA = "Exposure",
#'        ClassB = "Protein",LinkFrom = "default",LinkTo = "default")
#'    FuncExit(PID = res$PID)
#' @author Tianxiang Wu, Bin Wang (corresponding author)

EDBLink = function(PID,
                   OutPath = "default",
                   ClassA,
                   ClassB,
                   LinkFrom,
                   LinkTo){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'ExpoLink')
  res = httr::POST(url,
                   body = list(PID=PID,
                               ClassA =ClassA,
                               ClassB =ClassB,
                               LinkFrom =LinkFrom,
                               LinkTo =LinkTo
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/Link'))
  try(
    vroom::vroom_write(results$LinkTable,
                       paste0(OutPath,'/Link/Link_',ClassA, "_to_", ClassB,".csv"), delim = ","),
    silent = TRUE
  )
  return(results$LinkTable)
}

#' @title Initialize EBIO module
#' @description Initialize EBIO module analysis. It can generate an R6 class object
#'     integrating all the analysis information.
#' @usage InitEBIO()
#' @param
#' @details EBIO module is designed to find the biological relationships between
#'     exposure factors and health outcome. This module adopts the most frequently-used and
#'     authoritative databases, e.g., T3DB, CTD, ToxCast, StringDB, STITCH, KEGG, and GO.
#' @return An R6 class object.
#' @export
#' @examples res <- InitEBIO()
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Tianxiang Wu, Bin Wang,(corresponding author)


InitEBIO = function(){
  url = paste0(urlhead,'InitBioLink')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum=seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for EBIO module
#' @description Load data file for EBIO module.
#' @usage LoadEBIO(PID, UseExample = "default", DataPath=NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEBIO.
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input file directory, e.g. "D:/test/eg_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitEBIO()
#'    res = LoadEBIO(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Tianxiang Wu, Bin Wang,(corresponding author)

LoadEBIO =function(PID,
                   UseExample = "default",
                   DataPath = NULL){
  url = paste0(urlhead,'LoadBioLink')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
  }else{
    exdata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}

#' @title Convert different IDs to the unified ExposomeX IDs
#' @description Convert the IDs of exposure, chemicals, metabolites, or proteins to the unified ExposomeX ID,
#'     i.e., unified identifier in ExposomeX platform
#' @usage EBIOConvToExpoID(PID,OutPath)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEBIO.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @details
#' @return A data frame containing the converted ID information.
#' @export
#' @examples res = InitEBIO()
#'    res1 = LoadEBIO(PID = res$PID,UseExample = "example#1")
#'    res2 = EBIOConvToExpoID(PID = res$PID,OutPath = "default")
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Tianxiang Wu, Bin Wang,(corresponding author)

EBIOConvToExpoID = function(PID,
                            OutPath="default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'ConvToExpoID')
  res = httr::POST(url,
                   body = list(PID=PID),
                   encode = 'json')
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/EBIOConvToExpoID'))

  try(vroom::vroom_write(results$Data,paste0(OutPath,"/EBIOConvToExpoID/ExpoData.csv"),
                         delim = ","),silent = TRUE
  )
  return(results$Data)
}

#' @title Build the biological link
#' @description Build the biological link between the exposures and diseases
#' @usage EBIOBiolink(PID, OutPath="default",Mode, MetBiospec="blood")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEBIO.
#' @param OutPath chr. Output file directory. e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'    If "default", the current working directory will be set.
#' @param Mode chr. Method to build the biological link between exposures and diseases.
#'     Available options include "EPPD" (protein-protein interaction), "EGoD" (gene ontoloty),
#'     "EPD" (protein), "EGeD" (Gene), "EPaD" (pathway).
#' @param MetBiospec chr. Biological sample matrix for the metabolome analysis.
#'    Options include "blood" and "urine".
#' @details
#' @return A list object containing the edges and nodes of the biological link.
#' @export
#' @examples res = InitEBIO()
#'    res1 = LoadEBIO(PID = res$PID,UseExample = "example#1")
#'    res2 = EBIOConvToExpoID(PID = res$PID)
#'    res3 = EBIOBiolink(PID = res$PID, OutPath="default", Mode = "EGoD", MetBiospec = "blood")
#'    FuncExit(PID = res$PID)
#' @author Mingliang Fang, Tianxiang Wu, Bin Wang,(corresponding author)

EBIOBiolink = function(PID,
                       OutPath="default",
                       Mode,
                       MetBiospec = "blood"
){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'Biolink')
  res = httr::POST(url,
                   body = list(PID=PID,
                               Mode=Mode,
                               MetBiospec=MetBiospec
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/EBIOBiolink'))
  try(vroom::vroom_write(results$Edges,paste0(OutPath,"/EBIOBiolink/",Mode,"_edges.csv"),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$Nodes,paste0(OutPath,"/EBIOBiolink/",Mode,"_Nodes.csv"),
                         delim = ","),silent = TRUE
  )
  return(results)
}


#' @title Initialize EMETA Module
#' @description Initialize EMETA module, the first step to start EMETA Module.
#' @usage InitEMETA()
#' @details EMETA module mainly provides users preliminary information retrieval and screening for meta-analysis. Run InitEMETA to get a unique PID for following steps.
#' @return An R6 class object.
#' @export
#' @examples res = InitEMETA()
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

InitEMETA = function(){
  url = paste0(urlhead,'InitMeta')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum = seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load Data for EMETA Module
#' @description Upload local data file for EMETA Module.
#' @usage LoadEMETA(PID = NULL, UseExample = "default", DataPath = NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEMETA.
#' @param UseExample chr. A character indicates whether uses example data for analyses, available option include "example#1" for using example data1 and "default" for using data uploaded.
#' @param DataPath chr. Input file directory, e.g. "D:/test/eg_meta.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details After initializing EMETA module, the second step is to upload local data file for EMETA Module.
#'    Set param "UseExample = 'example#1'" to use example data1.
#'    LoadEMETA can only run successfully after successfully running InitEMETA.
#'    Please attention, PID must be got from the return result of InitEMETA().
#' @return An R6 class object containing the input data.
#' @export
#' @examples res = InitEMETA()
#'    res1 = LoadEMETA(PID = res$PID,
#'    UseExample = "example#1",
#'    DataPath = NULL)
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

LoadEMETA =function(PID = NULL,
                    UseExample = "default",
                    DataPath = NULL){
  url = paste0(urlhead,'LoadMeta')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)

  }else{
    exdata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Search or Download Articles' Main Information
#' @description Search or download articles' main information based on keywords.
#' @usage MetaRef(PID = NULL, Mode = "search", VarX, VarY, VarM, YearFrom, YearEnd, PMID, OutPath)
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEMETA.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default", the current working directory will be set. It should be noted that the slash symbol is "/", not "\".
#' @param Mode chr. Two modes are provided. “Search” for paper retrieval by keywords VarX/VarY/VarM/YearFrom/YearEnd, and “Download” for downloading main informain (main information only) for specified PMID.
#' @param VarX chr. “default” or a chemical names character (separate different values by ","). If “default”, the function will use the VarX values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the chemical names in the character instead.
#' @param VarY chr. “default” or a disease names character (separate different values by ","). If “default”, the function will use the VarY values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the disease names in the character instead.
#' @param VarM chr. “default” or a mediating factor names character (separate different values by ","). If “default”, the function will use the VarM values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the mediating factor names in the character instead.
#' @param YearFrom chr. “default” or a disease names character (separate different values by ","). Limits the time range searched for "search" mode. If “default”, the function will use the YearFrom values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the YearFrom in the character instead.
#' @param YearEnd chr. “default” or a disease names character (separate different values by ","). Limits the time range searched for "search" mode. If “default”, the function will use the YearEnd values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the YearEnd in the character instead.
#' @param PMID chr. “default” or a disease names character (separate different values by ","). Refs to the papers PMID for "download" mode. If “default”, the function will use the PMID values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the PMIDs in the character instead.
#' @details MetaRef provides the functions of paper retrieval and relevance sorting, returning the information to the user based on keywords.
#'    Please attention, PID must be got from the return result of InitEMETA().
#'    MetaRef can only run successfully after successfully running InitEMETA and LoadEMETA functions.
#' @return A list object containing dataframe of articles' information.
#' @examples res = InitEMETA()
#'    res1 = LoadEMETA(PID = res$PID,
#'    UseExample = "example#1",
#'    DataPath = NULL)
#'    res2 = MetaRef(PID = res$PID,
#'    OutPath = "default",
#'    Mode = "search",
#'    VarX = "default",
#'    VarY = "default",
#'    VarM = "default",
#'    YearFrom = "default",
#'    YearEnd = "default",
#'    PMID = "default")
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

MetaRef = function(PID = NULL,
                   Mode = "search",
                   VarX = "default",
                   VarY = "default",
                   VarM = "default",
                   YearFrom = "default",
                   YearEnd = "default",
                   PMID = "default",
                   OutPath = "default"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  url = paste0(urlhead,'MetaRef')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MetaReferMode=Mode,
                               MetaVarX=VarX,
                               MetaVarY=VarY,
                               MetaVarM=VarM,
                               YearFrom=YearFrom,
                               YearEnd=YearEnd,
                               PMID=PMID),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,"/Ref"))
  for (name in names(results$MetaRefer_Papers$`Meta-analysis`)){
    try(writexl::write_xlsx(results$MetaRefer_Papers$`Meta-analysis`[[name]],
                            paste0(OutPath,"/Ref/",name,"[meta].xlsx")),silent = TRUE)
  }
  for (name in names(results$MetaRefer_Papers$SystematicReview)){
    try(writexl::write_xlsx(results$MetaRefer_Papers$SystematicReview[[name]],
                            paste0(OutPath,"/Ref/",name,"[systematic review].xlsx.xlsx")),silent = TRUE)
  }
  for (name in names(results$MetaRefer_Papers$OtherPapers)){
    try(writexl::write_xlsx(results$MetaRefer_Papers$OtherPapers[[name]],
                            paste0(OutPath,"/Ref/",name,".xlsx")),silent = TRUE)
  }
  try(writexl::write_xlsx(results$MetaRefer_Papers$All,
                          paste0(OutPath,"/Ref/All_Papers.xlsx")),silent = TRUE)
  return(results)
}



#' @title Visualize the Articles' Main Information
#' @description Visualize the articles' main information after MetaRef function.
#' @usage VizRefer(PID = NULL, OutPath = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEMETA.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default", the current working directory will be set. It should be noted that the slash symbol is "/", not "\".
#' @details VizRefer provides the visual function of articles' main information, showing the year and region distribution directly.
#'    Please attention, PID must be got from the return result of InitEMETA().
#'    VizRefer can only run successfully after successfully running InitEMETA, LoadEMETA and MetaRef functions.
#' @return A list object containing plots of article information visualization.
#' @export
#' @examples res = InitEMETA()
#'   res1 = LoadEMETA(PID = res$PID,
#'   UseExample = "example#1",
#'   DataPath = NULL)
#'   res2 = MetaRef(PID = res$PID,
#'   OutPath = "default",
#'   Mode = "search",
#'   VarX = "default",
#'   VarY = "default",
#'   VarM = "default",
#'   YearFrom = "default",
#'   YearEnd = "default",
#'   PMID = "default")
#'   res3 = VizRefer(PID=res$PID,
#'   OutPath = "default")
#'   FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

VizRefer = function(PID = NULL,
                    OutPath = "default"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load('gridExtra')
  url = paste0(urlhead,'VizRefer')
  res = httr::POST(url,
                   body = list(PID=PID),
                   encode = 'json')
  results = unserialize(httr::content(res))
  for (name in names(results$MetaRefer_Viz)){
    try(ggplot2::ggsave(paste0(OutPath,'/Ref/',name,".png"),
                        plot=results$MetaRefer_Viz[[name]],
                        width = 35,height = 15,units = "cm"),silent = TRUE)
  }
  return(results)
}



#' @title Review Relationship Between Exposure and Outcome
#' @description Review the relationship between exposure and outcome besed on chemical ID and disease ID.
#' @usage MetaRev(PID = NULL, CID = "default", DID = "default", OutPath = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEMETA.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default", the current working directory will be set. It should be noted that the slash symbol is "/", not "\".
#' @param CID chr. “default” or a chemical ID character (separate different values by ","). If “default”, the function will use the Chemical_ID values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the chemical ID in the character instead. Chemical_ID refers to the target chemical ID which can be inchikey (eg. JIAARYAFYJHUJI-UHFFFAOYSA-L ), cas.rn (eg. 7784-42-1) or our EXC ID (eg. EX:C01631).
#' @param DID chr. “default” or a disease ID character (separate different values by ","). If “default”, the function will use the Disease_ID values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the disease ID in the character instead. Disease_ID refers to the target disease ID which can be MESH ID (format like MESH:D006973), OMIM ID (format like OMIM:182940) or our EXD ID (eg. EX:D16243)
#' @details MetaRev provides the functions of literature review. In the publishd papers, how many recorded that X is a protective/risky factor for Y? This questions can be solved by MetaRev function. (It can only search the papers available in our database)
#'    Please attention, PID must be got from the return result of InitEMETA().
#'    MetaRev can only run successfully after successfully running InitEMETA and LoadEMETA functions.
#' @return A list object containing relationship visualization.
#' @export
#' @examples res = InitEMETA()
#'    res1 = LoadEMETA(PID = res$PID,
#'    UseExample = "example#1",
#'    DataPath = NULL)
#'    res2 = MetaRev(PID = res$PID,
#'    OutPath = "default",
#'    CID = "default",
#'    DID = "default")
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

MetaRev = function(PID = NULL,
                   CID = "default",
                   DID = "default",
                   OutPath = "default"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load('gridExtra')
  url = paste0(urlhead,'MetaRev')
  res = httr::POST(url,
                   body = list(PID=PID,
                               Chemical_ID=CID,
                               Disease_ID =DID),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,"/Rev"))
  for (name in names(results$MetaReview_References)){
    try( writexl::write_xlsx(results$MetaReview_References[[name]],
                             paste0(OutPath,"/Rev/",name,".xlsx")),silent = TRUE
    )
  }

  for (name in names(results$MetaReview_Plot)){
    try(ggplot2::ggsave(paste0(OutPath,'/Rev/',name,'.jpg'),
                        plot = results$MetaReview_Plot[[name]],
                        width = 20,height = 15,units = "cm"),silent = TRUE)
  }
  return(results)
}



#' @title Pool Effect Value
#' @description Pool effect value in our meta database besed on chemical ID and disease ID.
#' @usage MetaAsso(PID = NULL, CID = "default", DID = "default", OutPath = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by InitEMETA.
#' @param OutPath chr. Output file directory, e.g. "D:/output". If "default", the current working directory will be set. It should be noted that the slash symbol is "/", not "\".
#' @param CID chr. “default” or a chemical ID character (separate different values by ","). If “default”, the function will use the Chemical_ID values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the chemical ID in the character instead. Chemical_ID refers to the target chemical ID which can be inchikey (eg. JIAARYAFYJHUJI-UHFFFAOYSA-L ), cas.rn (eg. 7784-42-1) or our EXC ID (eg. EX:C01631).
#' @param DID chr. “default” or a disease ID character (separate different values by ","). If “default”, the function will use the Disease_ID values in the file loaded by LoadEMETA. If a character (separate different values by ","), the function will use the disease ID in the character instead. Disease_ID refers to the target disease ID which can be MESH ID (format like MESH:D006973), OMIM ID (format like OMIM:182940) or our EXD ID (eg. EX:D16243)
#' @details MetaAsso provides the functions of effect value pooling. In the publishd papers, what is the effect value between X and Y? This questions can be solved by MetaAsso function. It can provide the combined results of fixed effect model and random effect model. (It can only search the papers available in EMETA database DB_Meta)
#'    Please attention, PID must be got from the return result of InitEMETA().
#'    MetaAsso can only run successfully after successfully running InitEMETA and LoadEMETA functions.
#' @return A list object containing forest plots.
#' @export
#' @examples res = InitEMETA()
#'    res1 = LoadEMETA(PID = res$PID,
#'    UseExample = "example#1",
#'    DataPath = NULL)
#'    res2 = MetaAsso(PID=res$PID,
#'    OutPath = "default",
#'    CID = "default",
#'    DID = "default")
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

MetaAsso = function(PID = NULL,
                    CID = "default",
                    DID = "default",
                    OutPath = "default"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load('gridExtra')
  url = paste0(urlhead,'MetaAsso')
  res = httr::POST(url,
                   body = list(PID=PID,
                               Chemical_ID=CID,
                               Disease_ID=DID),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,"/Asso"))
  for (name in names(results$MetaEffect_References)){
    try( writexl::write_xlsx(results$MetaEffect_References[[name]],
                             paste0(OutPath,"/Asso/",name,".xlsx")),silent = TRUE
    )
  }

  for (name in names(results$MetaEffect_Plot)){
    png(file = paste0(OutPath,'/Asso/',name,".png"), width = 4000,
        height = 80*results$MetaEffect_Plotheight[[name]]+500, res = 300)
    results$MetaEffect_Plot[[name]]-> plot
    print(plot)
    dev.off()
  }
  return(results)
}

#' @title  Initialize StatHEP module
#' @description Initialize StatHEP analysis. It can generate an R6 class object
#'     integrating all the analysis information.
#' @usage InitStatHEP()
#' @details Quantification of all trace organics in the biological fluids seems impossible and costly, regardless of the high individual exposure variability.
#' We hypothesized that the blood concentration of organic pollutants could be predicted via their exposure and chemical properties.
#' Developing a prediction model on the annotation of chemicals in human blood can provide new insight into the distribution and extent of exposures to a wide range of chemicals in humans.
#' Our objective of this module is to develop a machine learning model (eg., random forest) to predict blood concentrations of chemicals and prioritize chemicals of health concern.
#' Please see the details in Zhao et al., Environ Health Perspect. 2023 Mar;131(3):37009.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatHEP()
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

InitStatHEP = function(){
  url = paste0(urlhead,'InitHEP')
  seednum = sample(1000, 1)
  res = httr::POST(url,
                   body = list(seednum = seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}



#' @title  Load data file for StatHEP module
#' @description Upload data file for StatHEP module.
#' @usage LoadStatHEP(PID, UseExample= "default", DataPath)
#' @details
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatHEP.
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input file directory, e.g. "D:/test/eg_data.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatHEP()
#'     res <- LoadStatHEP(PID=res$PID, UseExample = "example#1", DataPath = NULL)
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

LoadStatHEP =function(PID,
                      UseExample="default",
                      DataPath=NULL){
  url = paste0(urlhead,'LoadHEP')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
  }else{
    exdata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title  Build StatHEP model
#' @description HEPPredBlood function is designed to integrate the multi-omic data to predict the incidence risk.
#'     It mainly aims to construct various stacked generalization models to predict the probability of outcome
#'     incidence, as well as providing the statistical explanation.
#' @usage HEPPredBlood(PID, OutPath, MC, N)
#' @details Users can easily get the modeling results and their visualization plots with high quality by following the detailed instructions in each step.
#' Machine learning model is used to predict blood concentrations of chemicals and prioritize chemicals of health concern.
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatHEP.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'    If "default", the current working directory will be set.
#' @param MC lgl. T (or TRUE) and F (or FALSE). Whether to perform Mentocaro simulation.
#' @param N num. Times of Mentocaro simulation.
#' @return A list containing StatHEP model results
#' @export
#' @examples res <- InitStatHEP()
#'    res1 <- LoadStatHEP(PID=res$PID, UseExample = "example#1", DataPath = NULL)
#'    res2 <- HEPPredBlood(PID=res$PID, OutPath = "default", MC = "T", N = 100)
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Bin Wang (corresponding author)

HEPPredBlood = function(PID,
                        OutPath = "default",
                        MC = F,
                        N = 100){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra)
  url = paste0(urlhead,'PredBlood')
  res = httr::POST(url,
                   body = list(PID = PID,
                               MC = MC,
                               N = N),
                   encode = 'json')

  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/HEPPredBlood'))
  try(vroom::vroom_write(results$eSet$model_cv_summary,
                         paste0(OutPath,'/HEPPredBlood/model_cv_summary.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$model_fit_data,
                         paste0(OutPath,'/HEPPredBlood/model_fit_data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$model_fit_rf_parameters,
                         paste0(OutPath,'/HEPPredBlood/model_fit_rf_parameters.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$model_fit_summary,
                         paste0(OutPath,'/HEPPredBlood/model_fit_summary.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$model_formulae,
                         paste0(OutPath,'/HEPPredBlood/model_formulae.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$pred_Cb,
                         paste0(OutPath,'/HEPPredBlood/pred_Cb.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$vars_importance,
                         paste0(OutPath,'/HEPPredBlood/vars_importance.csv'),
                         delim = ","),silent = TRUE
  )
  try(ggplot2::ggsave(plot=results$eSet$plot_model_fit,
                      filename = paste0(OutPath,'/HEPPredBlood/plot_model_fit.png'),
                      width = 10,
                      height = 7), silent = TRUE
  )
  return(results$eSet)
}



#' @title  Visualize StatHEP results
#' @description VizHEPPredBlood function is mainly aimed to visualize the modeling results calculated by HEPPredBlood function.
#'     It can provide plots with high quality of the final results to make it easier for users to understand.
#' @usage VizHEPPredBlood(PID, OutPath, MC,Layout,Brightness,Palette)
#' @details
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatHEP.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param MC lgl. T (or TRUE) and F (or FALSE). Whether to perform Mentocaro simulation.
#' @param Layout chr. Visualization layout. Available options include "forest" and "boxplot".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2" and other options about some journal preference styles
#'     including "cell", "nature", "science", "lancet", "nejm", etc.
#' @details You can get different styles of images by selecting different parameters.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatHEP()
#'    res1 <- LoadStatHEP(PID=res$PID, UseExample = "example#1", DataPath = NULL)
#'    res2 <- HEPPredBlood(PID=res$PID, OutPath = "default", MC = "T", N = 100)
#'    res3 <- VizHEPPredBlood(PID=res$PID, OutPath = "default", MC = "T",
#'    Layout = "forest", Brightness = "light", Palette = 'default1')
#'    FuncExit(PID = res$PID)
#' @author Weinan Lin, Ning Gao, Bin Wang (corresponding author)

VizHEPPredBlood = function(PID = res$PID,
                           OutPath = "default",
                           MC = F,
                           Layout = "forest",
                           Brightness = "light",
                           Palette = "default1"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra)
  url = paste0(urlhead,'VizPredBlood')
  res = httr::POST(url,
                   body = list(PID=PID,
                               MC = MC,
                               Layout=Layout,
                               Brightness=Brightness,
                               Palette=Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/VizHEPPredBlood'))
  try(ggplot2::ggsave(plot=results$eSet$plot_pred,
                      filename = paste0(OutPath,'/VizHEPPredBlood/plot_pred.png'),
                      width = 10,
                      height = 7), silent = TRUE)
  switch(Layout,
         "forest" = {
           try(ggplot2::ggsave(plot=results$eSet$HEPForestPlot[[stringr::str_c(Brightness,"_",Palette)]],
                               filename = paste0(OutPath,'/VizHEPPredBlood/HEPForestPlot.png'),
                               width = 10,
                               height = 15), silent = TRUE)
         },
         "boxplot" = {
           try(ggplot2::ggsave(plot=results$eSet$HEPboxplot[[stringr::str_c(Brightness,"_",Palette)]],
                               filename = paste0(OutPath,'/VizHEPPredBlood/HEPboxplot.png'),
                               width = 10,
                               height = 15), silent = TRUE)
         })
  return(results)
}


#' @title  Initialize StatMO module
#' @description Initialize StatMO analysis. It can generate an R6 class object
#'     integrating all the analysis information.
#' @usage InitStatMO()
#' @details StatMO module is designed to integrate the multi-omic data to predict the incidence risk.
#'     It mainly aims to construct various stacked generalization(SG) models to predict the probability
#'     of outcome incidence, as well as providing the statistical explanation. In addition, the module can provide
#'     visualization plots with high quality of the final calculation results to make it easier for users to understand.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatMO()
#'    FuncExit(PID = res$PID)
#' @author Guohuan Zhang, Yuting Wang, Bin Wang (corresponding author)

InitStatMO = function(){
  url = paste0(urlhead,'InitMO')
  seednum = sample(1000, 1)
  res = httr::POST(url,
                   body = list(seednum = seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  return(results$eSet)
}



#' @title  Load data file for multiomics module
#' @description Upload data file for multiomics module.
#' @usage LoadStatMO(PID, UseExample= "default", DataPath, VocaPath)
#' @details
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatMO.
#' @param UseExample chr. Method of uploading data. If "default", user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input file directory, e.g. "D:/test/eg_data.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input file vocabulary, e.g. "D:/test/eg_voca.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatMO()
#'     res1 <- LoadStatMO(PID=res$PID, UseExample = "example#1", DataPath = NULL, VocaPath = NULL)
#'     FuncExit(PID = res$PID)
#' @author Guohuan Zhang, Yuting Wang, Bin Wang (corresponding author)

LoadStatMO =function(PID,
                     UseExample="default",
                     DataPath= NULL,
                     VocaPath= NULL){
  url = paste0(urlhead,'LoadMO')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)
    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL
  }
  res = httr::POST(url,
                   body = serialize(list(PID=PID,
                                         UseExample=UseExample,
                                         Expodata=exdata,
                                         Vocadata=vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))

  return(results$eSet)
}


#' @title Transform data type
#' @description Transform data type
#' @usage TidyStatMO(PID,OutPath = "default",Vars,To)
#' @param PID chr. Program ID. It must be the same with the PID generated by initial functions.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Vars Variables to be imputed. Available options include:
#'     "all.x", all exposure variables; "all.c", all covariates;
#'     "all.cx", combination of All X and All C.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param To chr. Indicate the type of the chosen variables to be transformed into.
#'     Available options include "integer", "numeric", "character", "factor", "logical", and "date".
#' @details
#' @return An R6 class object containing the variable(s) after transforming data type.
#' @export
#' @examples res = InitStatMO()
#'     res1 <- LoadStatMO(PID=res$PID, UseExample = "example#1", DataPath = NULL, VocaPath = NULL)
#'     res1$Expo$Data
#'     res2 = TidyStatMO(PID=res$PID, Vars = "X1,X2", To = "factor")
#'     res2$Expo$Data
#'     FuncExit(PID = res$PID)
#' @author Guohuan Zhang, Yuting Wang, Bin Wang (corresponding author)

TidyStatMO = function(PID,
                      OutPath = "default",
                      Vars,
                      To){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'TidyMulOmics')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransTypeVars=Vars,
                               TransTypeTo=To),
                   encode = 'json')

  results = unserialize(httr::content(res))

  dir.create(paste0(OutPath,'/TidyStatMO'))
  try(vroom::vroom_write(results$eSet$Expo$Data,
                         paste0(OutPath,'/TidyStatMO/Expo_Data.csv'),
                         delim = ","),silent = TRUE
  )
  try(vroom::vroom_write(results$eSet$Expo$Voca,
                         paste0(OutPath,'/TidyStatMO/Expo_Voca.csv'),
                         delim = ","),silent = TRUE
  )

  return(results$eSet)
}

#' @title  Build multiomics model
#' @description MulOmicsCros function is designed to integrate the multi-omic data to predict the incidence risk.
#'     It mainly aims to construct various stacked generalization models to predict the probability of outcome
#'     incidence, as well as providing the statistical explanation.
#' @usage MulOmicsCros(PID, OutPath, OmicGroups,VarsY, VarsC, TuneMethod = "default", TuneNum, RsmpMethod, Folds, Ratio, Repeats, VarsImpThr, SG_Lrns)
#' @details The calculation time depends on the characteristics of your data, the number of learning methods,
#'     and the tuning method. For parameter "TuneMethod", the default option can provide faster calculations but less accurate
#'     results than other autotune methods. If you want to train a better model, choose other auto-tune method and increase the number of tuning times.
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatMO.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'    If "default", the current working directory will be set.
#' @param OmicGroups chr. Groups to be integrated. The groups of outcome and covariates or
#'     confounders are not included. Note that separates different learners by "," and without space(e.g. OmicGroups = "immunome,metabolome,proteome").
#' @param VarsY chr. Outcome variable for modelling. Only one variable can be entered.
#' @param VarsC chr. Covariates needing further statistical test. "all.c" option refers to all covariate variables listed in the data file.
#'     Users can also select part of them by copying available vars. Note that separates different vars by "," and without space(e.g. VarsC = "C1,C2").
#' @param TuneMethod chr. Method for hyper-parameter autotuning.
#'     Options include "default", "random_search", "grid_search", "nloptr"(Non-linear optimization), and "gensa"(Generalized simulated annealing).
#'     The "default" option uses the simple training method for parameter optimization of mlr3 package.
#' @param TuneNum num. Upper limit of model tuning times. It should be more than 20 times to search the appropriate parameters,
#'     but it takes more time. In theory, more time, better training results.
#' @param RsmpMethod chr. Method for resampling. Options include "cv"(cross validation), "loo"(leave-one-out cross validation),
#'     "bootstrap"(bootstrapping), "holdout"(holdout).
#' @param Folds num. Folds for cross validation resampling method. The default value is 5.
#' @param Ratio num. Ratio for "Holdout" resampling method. The default value is 5.
#' @param Repeats num. Repeats for "Bootstrap" resampling method.
#' @param VarsImpThr num. Threshold for feature selection.
#'     It refers to the ratio of accumulated importance of all variables of the selected variables for building the final model.
#' @param SG_Lrns chr. Learners for stacked generalization. Options include "lasso", "enet"(Elastic net), "rf"(Random forest), and "xgboost"(Xgboost).
#'     One or more arbitrary options can be selected at the same time. Note that separates different learners by "," and without space(e.g. SG_Lrns ="lasso,enet,rf,xgboost").
#' @return An R6 class object containing eight elements. The elements of that object include:
#' (1) "Importance": A list containing dataframes that contain the importance of features after modeling a single omic from different omicgroups.
#' (2) "Feature": A list containing dataframes that contain the the coefficients or importance of selected features after modeling a single omic from different learners.
#' (3) "Feature_select": A list containing dataframes that contain the selected features after modeling a single omic from different learners.
#' (4) "ModelStat": A list containing dataframes that contain the r-square value of the single omic model built by different learners.
#' (5) "Prediction_comp": A list containing dataframes that contain containing the prediction value of the SG model built by different combinations of learners.
#' (6) "SGModel_summary": A dataframe containing the r-square value of the SG model built by different combinations of learners.
#' (7) "NodeNum": The node number generated by different models.
#' (8) "SGplot": A visualized plot for SG model summary.
#' @export
#' @examples res <- InitStatMO()
#'    res1 <- LoadStatMO(PID=res$PID, UseExample = "example#1", DataPath = NULL, VocaPath = NULL)
#'    res2 <- TidyStatMO(PID=res$PID, Vars = "X1,X2", To = "factor")
#'    res3 <- MulOmicsCros(PID=res$PID, OutPath = "default", OmicGroups = "immunome,metabolome",
#'    VarsY = "Y1", VarsC = "all.c", TuneMethod = "random_search", TuneNum = 2, RsmpMethod = "cv", Folds = 2,
#'    Ratio = 0.67, Repeats = 5, VarsImpThr = 0.85, SG_Lrns ="lasso,enet")
#'    FuncExit(PID = res$PID)
#' @author Guohuan Zhang, Yuting Wang, Bin Wang (corresponding author)

MulOmicsCros = function(PID = res$PID,
                        OutPath = "default",
                        OmicGroups = "immunome,metabolome,proteome",
                        VarsY = "Y1",
                        VarsC = "all.c",
                        TuneMethod = "default",
                        TuneNum = 5,
                        RsmpMethod = "cv",
                        Folds = 5,
                        Ratio= 0.67,
                        Repeats= 5,
                        VarsImpThr = 0.85,
                        SG_Lrns ="lasso"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra,GGally)
  url = paste0(urlhead,'MulOmicsCros')
  res = httr::POST(url,
                   body = list(PID=PID,
                               OmicGroups=OmicGroups,
                               VarsY=VarsY,
                               VarsC=VarsC,
                               TuneMethod=TuneMethod,
                               TuneNum=TuneNum,
                               RsmpMethod=RsmpMethod,
                               Folds=Folds,
                               Ratio=Ratio,
                               Repeats=Repeats,
                               VarsImpThr=VarsImpThr,
                               SG_Lrns=SG_Lrns),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (OmicGroup in names(results$Importance)){

    dir.create(paste0(OutPath,"/1_",OmicGroup))
    dir.create(paste0(OutPath,"/1_",OmicGroup,"/Feature_Explain"))
    dir.create(paste0(OutPath,"/1_",OmicGroup,"/Features_Selection"))
    dir.create(paste0(OutPath,"/1_",OmicGroup,"/Model_Summary"))

    try(writexl::write_xlsx(results$Importance[[OmicGroup]]$lasso,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Feature_Explain/Importance_lasso.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Importance[[OmicGroup]]$enet,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Feature_Explain/Importance_enet.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Importance[[OmicGroup]]$rf,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Feature_Explain/Importance_rf.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Importance[[OmicGroup]]$xgboost,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Feature_Explain/Importance_xgboost.xlsx")),silent = TRUE)


    try(writexl::write_xlsx(results$Feature[[OmicGroup]]$lasso,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Coefficients_lasso.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature[[OmicGroup]]$enet,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Coefficients_enet.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature[[OmicGroup]]$rf,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Importance_rf.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature_select[[OmicGroup]]$xgboost,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Importance_xgboost.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature_select[[OmicGroup]]$lasso,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Selected.features_lasso.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature_select[[OmicGroup]]$enet,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Selected.features_enet.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature_select[[OmicGroup]]$rf,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Selected.features_rf.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Feature_select[[OmicGroup]]$xgboost,
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Features_Selection/Selected.features_xgboost.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$ModelStat[[OmicGroup]],
                            paste0(OutPath,"/1_",
                                   OmicGroup,
                                   "/Model_Summary/Measures_stat.xlsx")),silent = TRUE)
  }


  dir.create(paste0(OutPath,"/2_SG"))
  dir.create(paste0(OutPath,"/2_SG/Viz_Prediction"))
  dir.create(paste0(OutPath,"/2_SG/Model_Summary"))

  for (SG_Lrn in names(results$Prediction_comp)){
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$lasso$train,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by lasso Step2 by ",
                                   SG_Lrn,
                                   "_train.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$lasso$test,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by lasso Step2 by ",
                                   SG_Lrn,
                                   "_test.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$enet$train,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by enet Step2 by ",
                                   SG_Lrn,
                                   "_train.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$enet$test,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by enet Step2 by ",
                                   SG_Lrn,
                                   "_test.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$rf$train,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by rf Step2 by ",
                                   SG_Lrn,
                                   "_train.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$rf$test,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by rf Step2 by ",
                                   SG_Lrn,
                                   "_test.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$xgboost$train,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by xgboost Step2 by ",
                                   SG_Lrn,
                                   "_train.xlsx")),silent = TRUE)
    try(writexl::write_xlsx(results$Prediction_comp[[SG_Lrn]]$xgboost$test,
                            paste0(OutPath,"/",
                                   "2_SG/Viz_Prediction/",
                                   "Step1 by xgboost Step2 by ",
                                   SG_Lrn,
                                   "_test.xlsx")),silent = TRUE)
  }
  try(writexl::write_xlsx(results$SGModel_summary,
                          paste0(OutPath,"/",
                                 "2_SG/Model_Summary/",
                                 "Model assessment.xlsx")),silent = TRUE)
  try(writexl::write_xlsx(results$NodeNum,
                          paste0(OutPath,"/",
                                 "2_SG/Model_Summary/",
                                 "Node number.xlsx")),silent = TRUE)
  ggplot2::ggsave(plot=results$SGplot,
                  filename = paste0(OutPath,"/2_SG/Model_Summary/Model assessment_boxplot.png"),
                  width = 10,
                  height = 3.5*results$plotheight)
  return(results)
}




#' @title  Visualize multiomics model results
#' @description VizMultOmic function is mainly aimed to visualize the modeling results calculated by MulOmicsCros function.
#'     It can provide plots with high quality of the final results to make it easier for users to understand.
#' @usage VizMultOmic(PID, OutPath, VarsY, NodeNum, EdgeThr, Layout, Brightness, Palette)
#' @details
#' @param PID chr. Program ID. It must be the same with the PID generated by InitStatMO.
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that the slash symbol is "/", not "\".
#'     If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable for modeling. Only one variable can be entered.
#' @param NodeNum num. Number of nodes in the network plot. The maximum number is generated by Multiomics function.
#'     User can set a smaller value if needed.
#' @param EdgeThr num. Threshold of correlation coefficient ranging 0-1 for generating the concerned edges of the network plot.
#' @param Layout chr. Visualization layout. Available options include "force-directed" and "degree-circle".
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2" and other options about some journal preference styles
#'     including "cell", "nature", "science", "lancet", "nejm", etc.
#' @details You can get different styles of images by selecting different parameters.
#' @return An R6 class object containing seven elements. The elements of that object include:
#' (1) "Importance_plot": Plots for the importance of features after modeling a single omic from different learners.
#' (2) "Measures_boxplot": Plots for the r-square value of the single omic model built by different learners.
#' (3) "NetWork_State": A list containing dataframes that contain nodes and edges used to draw
#'     interOmic network from models built by different learners.
#' (4) "Nodeplot": Interomic node plots from models built by different models.
#' (5) "Networkplot": Interomic network plots from models built by different models.
#' (6) "Prediction_train_plot": Plots for the prediction value of the train set in the SG model built by different combinations of learners.
#' (7) "Prediction_test_plot": Plots for the prediction value of the test set in the SG model built by different combinations of learners.
#' @export
#' @examples res <- InitStatMO()
#'    res1 <- LoadStatMO(PID=res$PID, UseExample = "example#1", DataPath = NULL, VocaPath = NULL)
#'    res2 <- TidyStatMO(PID=res$PID, Vars = "X1,X2", To = "factor")
#'    res3 <- MulOmicsCros(PID=res$PID, OutPath = "default", OmicGroups = "immunome,metabolome",
#'    VarsY = "Y1", VarsC = "all.c", TuneMethod = "random_search", TuneNum = 2, RsmpMethod = "cv", Folds = 2,
#'    Ratio = 0.67, Repeats = 5, VarsImpThr = 0.85, SG_Lrns ="lasso,enet")
#'    res4 <- VizMultOmic(PID=res$PID, OutPath = "default", VarsY = "Y1", NodeNum=100, EdgeThr= 0.45,
#'    Layout = "force-directed", Brightness = "light", Palette = 'default1')
#'    FuncExit(PID = res$PID)
#' @author Guohuan Zhang, Yuting Wang, Ning Gao, Bin Wang (corresponding author)

VizMultOmic = function(PID = res$PID,
                       OutPath = "default",
                       VarsY = "Y1",
                       NodeNum,
                       EdgeThr,
                       Layout = "force-directed",
                       Brightness = "light",
                       Palette = "default1"){
  if(OutPath == 'default'){
    OutPath = getwd()
  }
  pacman::p_load(gridExtra,GGally)
  url = paste0(urlhead,'VizMulOmicCros')
  res = httr::POST(url,
                   body = list(PID=PID,
                               VarsY=VarsY,
                               NodeNum = NodeNum,
                               EdgeThr = EdgeThr,
                               Layout=Layout,
                               Brightness=Brightness,
                               Palette=Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (OmicGroup in names(results$Importance_plot)){
    for (learner in names(results$Importance_plot[[OmicGroup]])){
      try(ggplot2::ggsave(
        plot = results$Importance_plot[[OmicGroup]][[learner]],
        filename = paste0(OutPath,"/1_",OmicGroup,
                          "/Feature_Explain/Importance_",learner,"_plot.png"),
        width = 6,
        height = 7),silent = TRUE)

    }

    try(ggplot2::ggsave(plot=results$Measures_boxplot[[OmicGroup]],
                        filename = paste0(OutPath,"/1_",OmicGroup,
                                          "/Model_Summary/Measures_boxplot.png"),
                        width = 10,
                        height = 3.5),
        silent = TRUE

    )
  }

  dir.create(paste0(OutPath,"/2_SG/Viz_InterOmic"))
  for (learner in names(results$NetWork_State)){
    try(writexl::write_xlsx(results$NetWork_State[[learner]]$node,
                            paste0(OutPath,"/2_SG/Viz_InterOmic/Node_Step1.by.",learner,".xlsx")),silent = TRUE)

    try(writexl::write_xlsx(results$NetWork_State[[learner]]$edge,
                            paste0(OutPath,"/2_SG/Viz_InterOmic/Edge_Step1.by.",learner,".xlsx")),silent = TRUE)

  }
  for ( learner in names(results$Nodeplot)){
    try(ggplot2::ggsave(filename = paste0(OutPath,"/2_SG/Viz_InterOmic/NodePlot_",learner,"_",
                                          Brightness,"_",Palette,".png"),
                        plot = results$Nodeplot[[learner]],
                        height = 6,
                        width = 8),
        silent = TRUE
    )
    try(ggplot2::ggsave(paste0(OutPath,"/2_SG/Viz_InterOmic/NetworkPlot_",learner,"_",
                               Layout,"_",Brightness,"_",Palette,".png"),
                        plot = results$Networkplot[[learner]],
                        width =  20,
                        height = 20,
                        units = "cm"),
        silent = TRUE
    )
  }

  for (learner1 in names(results$Prediction_train_plot)){
    for (learner2 in names(results$Prediction_train_plot[[learner1]])){
      try(
        ggplot2::ggsave(plot=results$Prediction_train_plot[[learner1]][[learner2]],
                        filename = paste0(OutPath,"/2_SG/Viz_Prediction/",
                                          learner1," ",learner2,"_Training.png"),
                        width = 6,
                        height = 6),
        silent = TRUE
      )
      try(
        ggplot2::ggsave(plot=results$Prediction_test_plot[[learner1]][[learner2]],
                        filename = paste0(OutPath,"/2_SG/Viz_Prediction/",
                                          learner1," ",learner2,"_Test.png"),
                        width = 6,
                        height = 6),
        silent = TRUE
      )
    }
  }
  return(results)
}

#' @title Initialize StatMix module
#' @description Initialize StatMix module analysis. It can generate an R6 class object.
#' @usage InitStatMix()
#' @param
#' @details StatMix module is designed to analyze mixture effect of the various
#'     exposure factors. It mainly aims to screen the representative features with
#'     high contribution to the health outcome, as well as their potential interaction effect.
#' @return An R6 class object.
#' @export
#' @examples res <- InitStatMix()
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

InitStatMix = function(){
  url = paste0(urlhead,'InitMix')
  seednum = sample(1000,1)
  res = httr::POST(url,
                   body = list(seednum = seednum),
                   encode = 'json')
  results = unserialize(httr::content(res))
  results = unserialize(httr::content(res))
  return(results$eSet)
}


#' @title Load data file for StatMix module
#' @description Load data file for StatMix module
#' @usage LoadStatMix(PID, UseExample = "default", DataPath = NULL, VocaPath = NULL)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param UseExample chr. Method of uploading data. If "default",user should upload their own data files,
#'    or use "example#1" provided by this module.
#' @param DataPath chr. Input directory of data file, e.g. "D:/test/eg_data_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @param VocaPath chr. Input directory of vocabulary file, e.g. "D:/test/eg_voca_biolink.xlsx". It should be
#'    noted that the slash symbol is "/", not "\".
#' @details
#' @return An R6 class object containing the input data.
#' @export
#' @examples res <- InitStatMix()
#'    res = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

LoadStatMix = function(PID,
                       UseExample = "default",
                       DataPath = NULL,
                       VocaPath = NULL){
  url = paste0(urlhead,'LoadMix')
  if(UseExample=="default"){
    exdata = readxl::read_excel(DataPath)

    vodata = readxl::read_excel(VocaPath)
  }else{
    exdata = NULL
    vodata = NULL}
  res = httr::POST(url,
                   body = serialize(list(PID = PID,
                                         UseExample = UseExample,
                                         Expodata = exdata,
                                         Vocadata = vodata),connection = NULL),
                   httr::content_type('application/x-rds'),
                   encode = 'raw')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Tidy data for Mix module
#' @description Tidy data file for Mix module
#' @usage TidyStatMix(PID = PID, TransDummyVars = ...)
#' @param PID chr. Program ID. It must be the same with the PID generated by
#'    InitStatMix.
#' @param TransDummyVars chr. A vector indicating factor variables to be transformed
#'    as dummy variables. Default value is "default" where all factor vars will be
#'    converted into dummies.
#' @param LogTransM chr. Method of converting data distribution. Available options
#'    include "log10"(default), "log2", and "ln".
#' @param RangeLow num. Lower range value for data scaling. Default value is 0.
#'    RangeLow value should be lower than that of RangeUpp.
#' @param RangeUpp num. Upper range value for data scaling. Default value is 1.
#'    RangeUpp value should be larger than that of RangeLow.
#' @details TidyStatMix function resembles several tidy steps for data pre-treatment
#'    before executing other functions in Mix module, including deleting vars with
#'    low variance, transforming categorical vars into factors, transferring data
#'    distribution and scaling for exposures.
#' @return An R6 class object containing the input data and vocabulary file.
#' @export
#' @examples res <- InitStatMix()
#'    res1 <- LoadStatMix(PID = res$PID, UseExample = "example#1", DataPath = NULL,
#'    VocaPath = NULL)
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

TidyStatMix =function(PID,
                      TransDummyVars = "default",
                      LogTransM="log10",
                      RangeLow=RangeLow,
                      RangeUpp=RangeUpp){
  url = paste0(urlhead,'TidyMix')
  res = httr::POST(url,
                   body = list(PID=PID,
                               TransDummyVars=TransDummyVars,
                               LogTransM=LogTransM,
                               RangeLow=RangeLow,
                               RangeUpp=RangeUpp),
                   encode = 'json')
  results = unserialize(httr::content(res))
  # results list
  return(results$eSet)
}


#' @title Build multiple linear regression (MLR) model
#' @description Build multiple linear regression (MLR) model
#' @usage MixMLR(PID,OutPath = "default",VarsY,VarsX,IncCova = "F",
#'     SelMethod = "lasso",PredType = "response",Family)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available
#'     variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param Covariates chr. To determine covariates in models. Default is "all.c", where all
#'     candidate covariates will be included in models. Users can also choose available
#'     variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "C1,C2,C3".
#' @param SelMethod chr. Method to select the important features to the final model.
#'     Options include "stepwise" (stepwise regression), "lasso" (Regularization regression of
#'     least absolute shrinkage and selection operator), and "enet" (Regularization regression of elastic net).
#' @param PredType chr. Prediction type of the outcome variable, including "response"
#'     for the actual values and "prob" for outcome with binary variable.
#' @param Family chr. The link function for the regression model according the data type of outcomes,
#'     including "gaussian" for continuous variable, "binomial" for binary variable,
#'     and "poisson" for counting variable
#' @details
#' @return A list containing the MLR analysis.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = res3 = MixMLR(PID = res$PID, VarsY = "Y1", VarsX = "all.x",
#'    Covariates = "all.c", SelMethod = "lasso", PredType = "response",
#'    Family = "gaussian")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

MixMLR = function(PID,
                  OutPath = "default",
                  VarsY,
                  VarsX = "all.x",
                  Covariates = "all.c",
                  SelMethod = "lasso",
                  PredType = "response",
                  Family){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'MixMLR')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsX = VarsX,
                               Covariates = Covariates,
                               SelMethod = SelMethod,
                               PredType = PredType,
                               Family = Family),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/MixMLR'))
  for(name in names(results$MLRTable)){
    try(
      vroom::vroom_write(results$MLRTable[[name]],
                         paste0(OutPath,'/MixMLR/', name,".csv"),
                         delim = ","),silent = TRUE
    )
  }

  return(results$MLRTable)
}



#' @title Visualize the model results
#' @description Visualize the results of multiple linear regression (MLR) model
#' @usage VizMixMLR(PID,OutPath = "default",VarsY,SelMethod,Brightness = "light",Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param SelMethod chr. Method to select the important features to the final model.
#'     Options include "stepwise" (stepwise regression), "lasso" (Regularization regression of
#'     least absolute shrinkage and selection operator), and "enet" (Regularization regression of elastic net).
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing the MLR analysis plot.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = res3 = MixMLR(PID = res$PID, VarsY = "Y1", VarsX = "all.x",
#'    Covariates = "all.c", SelMethod = "lasso", PredType = "response",
#'    Family = "gaussian")
#'    res4 = VizMixMLR(PID=res$PID, VarsY = "Y1", SelMethod = 'lasso',
#'    Brightness = "light", Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

VizMixMLR = function(PID,
                     OutPath = "default",
                     VarsY,
                     SelMethod,
                     Brightness = "light",
                     Palette = "default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'VizMixMLR')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VizVarsY = VarsY,
                               SelMethod = SelMethod,
                               Brightness = Brightness,
                               Palette = Palette),
                   encode = 'json',
                   httr::timeout(600))
  results = unserialize(httr::content(res))

  for (name in names(results$MLRPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/MixMLR/',name,".png"),
                      plot = results$MLRPlot[[name]],
                      width =  results$MLRPlot_para[[name]][['width']],
                      height = results$MLRPlot_para[[name]][['height']],
                      units = "cm",dpi = 300),silent = TRUE
    )
  }
  return(results$MLRPlot)
}

#' @title Build g-computation model
#' @description Build g-computation model
#' @usage MixGcomp(PID, VarsY, VarsX, Covariates = "all.c", Family, q = 10)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3".
#' @param Covariates chr. To determine covariates in models. Default is "all.c", where all
#'     candidate covariates will be included in models. Users can also choose available
#'     variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "C1,C2,C3".
#' @param Family chr. The link function for the regression model according the data type of outcomes,
#'     including "gaussian" for continuous variable, "binomial" for binary variable,
#'     and "poisson" for counting variable.
#' @param q num. Levels for ranking mixture variables, e.g. in quartiles (q = 4), deciles (q = 10),
#'     or percentiles (q = 100). The default is 10.
#' @details
#' @return A list containing the WQS analysis results.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = MixGcomp(PID=res$PID, VarsY = "Y1", VarsX = "all.x",
#'    Covariates = "all.c", Family = "gaussian", q = 10)
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

MixGcomp = function(PID,
                    OutPath = "default",
                    VarsY,
                    VarsX = "all.x",
                    Covariates = "all.c",
                    Family,
                    q = 10){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'MixGcomp')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsX = VarsX,
                               Covariates = Covariates,
                               Family = Family,
                               q = q
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/MixGcomp'))
  for(name in names(results$GcompTable)){
    try(
      vroom::vroom_write(results$GcompTable[[name]],
                         paste0(OutPath,'/MixGcomp/', name,".csv"),
                         delim = ","),silent = TRUE
    )
  }
  return(results$GcompTable)
}


#' @title Visualize results of g-computation model
#' @description Visualize results of g-computation model
#' @usage VizMixGcomp(PID, OutPath = "default", Brightness  = "dark", Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing the g-computation analysis results' plot.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = MixGcomp(PID=res$PID, VarsY = "Y1", VarsX = "all.x",
#'    Covariates = "all.c", Family = "gaussian", q = 10)
#'    res4 = VizMixGcomp(PID = res$PID, VarsY = "Y1", Brightness = "dark",
#'    Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

VizMixGcomp = function(PID,
                       VarsY,
                       OutPath = "default",
                       Brightness  = "dark",
                       Palette = "default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'VizMixGcomp')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               Brightness = Brightness,
                               Palette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (name in names(results$GcompPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/MixGcomp/',name,".png"),
                      plot = results$GcompPlot[[name]],
                      width = 35,
                      height = 20,
                      units = "cm",dpi = 300),silent = TRUE
    )
  }
  return(results$GcompPlot)
}



#' @title Build weighted quantile sum regression (WQS) model
#' @description Build weighted quantile sum regression (WQS) model
#' @usage MixWQS(PID, OutPath = "default", VarsY, VarsX, IncCova = "F", Family,
#'     VarStrat = "none",RatioValidat = 0.3,q = 10,b = 100,b1_pos = F,b1_constr = F)
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param Covariates chr. To determine covariates in models. Default is "all.c", where all
#'     candidate covariates will be included in models. Users can also choose available
#'     variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "C1,C2,C3".
#' @param Family chr. The link function for the regression model according the data type of outcomes,
#'     including "gaussian" for continuous variable, "binomial" for binary variable,
#'     and "poisson" for counting variable.
#' @param VarStrat chr. A factor variable used for stratifying for the model.
#' @param RatioValidat num. Percentage of the dataset to be used to validate the model.
#'     If validation = 0 then the test dataset is used as validation dataset too. The default is 0.3.
#' @param q num. Levels for ranking mixture variables, e.g. in quartiles (q = 4), deciles (q = 10),
#'     or percentiles (q = 100). The default is 10.
#' @param b num. Number of bootstrap samples used in parameter estimation. The default is 100.
#' @param b1_pos lgl. T (or TRUE) and F (or FALSE). Whether the beta values were positive
#'     to derive weights from to build models.
#' @param b1_constr lgl. T (or TRUE) and F (or FALSE). A logial value that determines
#'     whether to apply positive (if b1_pos = TRUE) or negative (if b1_pos = FALSE) constraints
#'     in the optimization function for the weight estimation.
#' @details
#' @return A list containing the WQS analysis results.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = MixWQS(PID=res$PID, VarsY = "Y1", VarsX = "all.x",Covariates = "all.c",
#'    Family = "gaussian", VarStrat = "none", RatioValidat = 0.3, q = 10, b=100,
#'    b1_pos = 'F', b1_constr = 'F')
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

MixWQS = function(PID,
                  OutPath = "default",
                  VarsY,
                  VarsX,
                  Covariates = "all.c",
                  Family,
                  VarStrat = "none",
                  RatioValidat = 0.3,
                  q = 10,
                  b = 100,
                  b1_pos = F,
                  b1_constr = F){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'MixWQS')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsX = VarsX,
                               Covariates = Covariates,
                               Family = Family,
                               VarStrat = VarStrat,
                               RatioValidat = RatioValidat,
                               q = q,
                               b = b,
                               b1_pos = b1_pos,
                               b1_constr = b1_constr
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/MixWQS'))
  for(name in names(results$WQSTable)){
    try(
      vroom::vroom_write(results$WQSTable[[name]],
                         paste0(OutPath,'/MixWQS/', name,".csv"),
                         delim = ","),silent = TRUE
    )
  }
  return(results$WQSTable)
}


#' @title Visualize results of weighted quantile sum regression (WQS) model
#' @description Visualize results of weighted quantile sum regression (WQS) model
#' @usage VizMixWQS(PID, OutPath = "default", VarsY, Brightness  = "dark", Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing the WQS analysis results' plot.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = MixWQS(PID=res$PID, VarsY = "Y1", VarsX = "all.x",Covariates = "all.c",
#'    Family = "gaussian", VarStrat = "none", RatioValidat = 0.3, q = 10, b=100,
#'    b1_pos = 'F', b1_constr = 'F')
#'    res4 = VizMixWQS(PID = res$PID, VarsY = "Y1", Brightness = "dark",
#'    Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

VizMixWQS = function(PID,
                     OutPath = "default",
                     VarsY,
                     Brightness  = "dark",
                     Palette = "default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'VizMixWQS')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VizVarsY = VarsY,
                               Brightness = Brightness,
                               Palette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (name in names(results$WQSPlot)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/MixWQS/',name,".png"),
                      plot = results$WQSPlot[[name]],
                      width = 35,
                      height = 20,
                      units = "cm",dpi = 300),silent = TRUE
    )
  }
  return(results$WQSPlot)
}


#' @title Build the Bayesian Kernel Machine Regression (BKMR) model
#' @description Build the Bayesian Kernel Machine Regression (BKMR) model
#' @usage MixBKMR(PID, OutPath = "default", VarsY, VarsX, IncCova,
#'     Family, Group = F,Iter = 2000, qfixed = 0.5, qsbivar = "default",
#'     qsoverall = "default", qsdiff = "default")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param VarsX chr. Exposure variable used for modeling. The default option is
#'     "all.x" (All exposure variables are included). Users can also choose available variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "X1,X2,X3"
#' @param Covariates chr. To determine covariates in models. Default is "all.c", where all
#'     candidate covariates will be included in models. Users can also choose available
#'     variables.
#'     It should be noted that there is fixed format for the entering characters
#'     separated with comma and without space, e.g., "C1,C2,C3".
#' @param Family chr. The link function for the regression model according the data type of outcomes,
#'     including "gaussian" for continuous variable, "binomial" for binary variable,
#'     and "poisson" for counting variable
#' @param Group lgl. T (or TRUE) and F (or FALSE). Whether to use group indicators
#'     for fitting hierarchical variable selection. If "TRUE", the group name (GroupName)
#'     should be provided in the vocabulary data file.
#' @param Iter num. Number of iterations for modeling. The default is 500.
#'     For more accurate and stable results, a minimum of 10,000 iteration is recommended.
#' @param qfixed num. Quantile at which to fix the other predictors. The default is 0.5.
#' @param qsbivar chr. Quantiles at which to fix the second variable. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., the default character sequence is "0.1,0.5,0.9"
#' @param qsoverall num. Quantiles at which to calculate the overall risk summary. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., the default character sequence is "0.25,0.30,0.35,0.40,0.45,0.50,0.55,0.60,0.65,0.70,0.75"
#' @param qsdiff chr. Indicating the two quantiles for computing their effect difference. It should be noted that
#'     there is fixed format for the entering characters separated with comma and without space,
#'     e.g., the default character sequence is "0.25,0.75".
#' @details
#' @return A list containing the BKMR analysis results.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = MixBKMR(PID = res$PID, VarsY = "Y1", VarsX = "X4,X5,X6,X7,X8,X9,X10",
#'    Covariates = "all.c", Family = "gaussian", Group = 'F', Iter = 2000, qfixed = 0.5,
#'    qsbivar = "default", qsoverall = "default", qsdiff = "default")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

MixBKMR = function(PID,
                   OutPath = "default",
                   VarsY,
                   VarsX = "all.x",
                   Covariates = "all.c",
                   Family,
                   Group = F,
                   Iter = 2000,
                   qfixed = 0.5,
                   qsbivar = "default",
                   qsoverall = "default" ,
                   qsdiff = "default"){
  if(OutPath == "default"){
    OutPath = getwd()
  }
  url = paste0(urlhead,'MixBKMR')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VarsY = VarsY,
                               VarsX = VarsX,
                               Covariates = Covariates,
                               Family = Family,
                               Group = Group,
                               Iter = Iter,
                               qfixed = qfixed,
                               qsbivar = qsbivar,
                               qsoverall = qsoverall,
                               qsdiff = qsdiff
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))
  dir.create(paste0(OutPath,'/MixBKMR'))
  for(name in names(results$TableBKMR)){
    try(
      vroom::vroom_write(results$TableBKMR[[name]],
                         paste0(OutPath,'/MixBKMR/', name,".csv"),
                         delim = ","),silent = TRUE
    )
  }
  return(results$TableBKMR)
}


#' @title Visualize the model results of Bayesian Kernel Machine Regression (BKMR)
#' @description Visualize the model results of Bayesian Kernel Machine Regression (BKMR)
#' @usage VizMixBKMR(PID, OutPath = "default", VarsY, Brightness = "dark", Palette = "default1")
#' @param PID chr. Program ID. It must be the same with the PID generated by StatMix
#' @param OutPath chr. Output file directory, e.g. "D:/test". It should be noted that
#'     the slash symbol is "/", not "\". If "default", the current working directory will be set.
#' @param VarsY chr. Outcome variable used for modeling. Only one variable can be entered.
#' @param Brightness chr. Visualization brightness. Available options include "light" and "dark".
#' @param Palette chr. Visualization palette. Available options include "default1", "default2"
#'    and several journal preference styles (i.e., cell, nature, science, lancet, nejm, and jama).
#' @details
#' @return A list containing the BKMR analysis results' plot.
#' @export
#' @examples res <- InitStatMix()
#'    res1 = LoadStatMix(PID = res$PID, UseExample = "example#1")
#'    res2 <- TidyStatMix(PID = res$PID, LogTransM = "log10", RangeLow = 0, RangeUpp = 1)
#'    res3 = MixBKMR(PID = res$PID, VarsY = "Y1", VarsX = "X4,X5,X6,X7,X8,X9,X10",
#'    Covariates = "all.c", Family = "gaussian", Group = 'F', Iter = 2000, qfixed = 0.5,
#'    qsbivar = "default", qsoverall = "default", qsdiff = "default")
#'    res4 = VizMixBKMR(PID=res$PID, VarsY = "Y1", Brightness = "dark",
#'    Palette = "default1")
#'    FuncExit(PID = res$PID)
#' @author Mengyuan Ren, Bin Wang(corresponding author)

VizMixBKMR = function(PID,
                      OutPath = "default",
                      VarsY,
                      Brightness = "dark",
                      Palette = "default1"){
  if(OutPath == "default"){
    OutPath = getwd()
  }

  message("It may take some time. Thanks for your patient waiting!")

  url = paste0(urlhead,'VizMixBKMR')
  res = httr::POST(url,
                   body = list(PID = PID,
                               VizVarsY = VarsY,
                               Brightness = Brightness,
                               Palette = Palette
                   ),
                   encode = 'json')
  results = unserialize(httr::content(res))

  for (name in names(results$BKMRPlot$plot_pip.univar)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/MixBKMR/plot_pip_univar',name,".png"),
                      plot = results$BKMRPlot$plot_pip.univar[[name]],
                      width =  results$BKMRPlot_para$plot_pip.univar[[name]][['width']],
                      height = results$BKMRPlot_para$plot_pip.univar[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  for (name in names(results$BKMRPlot$plot_pred.resp.bivar.all)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/MixBKMR/plot_pred_resp_bivar_all',name,".png"),
                      plot = results$BKMRPlot$plot_pred.resp.bivar.all[[name]],
                      width =  results$BKMRPlot_para$plot_pred.resp.bivar.all[[name]][['width']],
                      height = results$BKMRPlot_para$plot_pred.resp.bivar.all[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  for (name in names(results$BKMRPlot$plot_risks.overall)){
    try(
      ggplot2::ggsave(paste0(OutPath,'/MixBKMR/plot_risks_overall',name,".png"),
                      plot = results$BKMRPlot$plot_risks.overall[[name]],
                      width =  results$BKMRPlot_para$plot_risks.overall[[name]][['width']],
                      height = results$BKMRPlot_para$plot_risks.overall[[name]][['height']],
                      units = "cm"),silent = TRUE
    )
  }
  return(results)
}



#' @title End the module analysis
#' @description End the module analysis
#' @usage FuncExit(PID)
#' @param PID chr. Program ID. It should be the same with the PID generated by Init.
#' @details
#' @return Exit status
#' @export
#' @examples res = InitEBIO()
#'    FuncExit(PID = res$PID)
#' @author Bin Wang (corresponding author)

FuncExit = function(PID){
  url = paste0(urlhead,'Exit')
  res = httr::POST(url,
                   body = list(PID=PID),
                   encode = 'json')
  if(res$status_code == 200){
    return("Success to exit. Thanks for using ExposomeX platform!")
  }else{
    return("No process needs exit. Thanks for using ExposomeX platform!")
  }
}
